---
title: "Testing"
author: "Deepak Tanwar"
date: "<b>Created on:</b> 2020-09-10 <br> <b>Updated on:</b> `r Sys.Date()`"
output:
  rmdformats::readthedown:
    highlight: tango
    self_contained: true
    thumbnails: true
    lightbox: true
    gallery: false
    keep_md: no
    number_sections: no
    fig_width: 8
    fig_height: 8
    fig_caption: true
    df_print: paged
    code_folding: show
  fontsize: 12pt
  geometry: margin=1in
  documentclass: article
# bibliography: references.bib
link-citations: yes
---


# Library
```{r overlap-matrix-rGREAT-1, warning = F, message = F}
library(ggplot2)
library(readxl)
library(patchwork)
library(dplyr)
```

# ATAC-Seq

Here are the results from differentially accessibility analysis. We are selecting the regions which have distance to TSS <= 1kbp.

```{r}
da <- data.frame(read_xlsx("input/atac/atac_diff_chromatin_accessibility_lfc1_qval05.xlsx"))
da <- da[abs(da$Peak.distance2nearestTSS_pa) <= 1000, c(6, 24, 25, 29, 33)]
colnames(da) <- c("Peak_name", "Peak_distance2TSS", "Genes", "DA_logFC", "DA_qvalue")
dim(da)
```


# RNA-Seq (lab)

Here are the results from differentially expression analysis from lab's data.

```{r}
load("input/rnaseq/limma_SC_Controls.RData")
dea_lab <- dea.limma$`PND15 vs Adult`[, c(1, 6, 10)]
colnames(dea_lab) <- c("Genes", "DEA_lab_logFC", "DEA_lab_qvalue")

dea_lab_sig <- dea_lab[abs(dea_lab$DEA_lab_logFC) >= 1 & dea_lab$DEA_lab_qvalue <= 0.05, ]
```


# RNA-Seq (lit)

Here are the results from differentially expression analysis from data from literature.

```{r}
load("input/rna_lit/limma_SC_Controls_lit.RData")
dea_lit <- dea.limma$`PND14 vs Adult`[, c(1, 2, 6)]
colnames(dea_lit) <- c("Genes", "DEA_lit_logFC", "DEA_lit_qvalue")

dea_lit_sig <- dea_lit[abs(dea_lit$DEA_lit_logFC) >= 1 & dea_lit$DEA_lit_qvalue <= 0.05, ]
```



# Merged table

Merging all tables with gene symbols

## Merged with RNA-Seq
```{r}
tab <- Reduce(merge, list(da, dea_lab, dea_lit))
head(tab)

dim(tab)
```

## Changed in same direction from both RNA-Seq
```{r}
tab_dir <- tab[(tab$DEA_lab_logFC > 0 & tab$DEA_lit_logFC > 0) | (tab$DEA_lab_logFC < 0 & tab$DEA_lit_logFC < 0),]
```


## Merged with significant RNA-Seq (Intersection)
```{r}
tab_sig <- Reduce(merge, list(da, dea_lab_sig, dea_lit_sig))
head(tab_sig)

dim(tab_sig)
```

## Changed in same direction from both RNA-Seq (significant)
```{r}
tab_dir_sig <- tab_sig[(tab_sig$DEA_lab_logFC > 0 & tab_sig$DEA_lit_logFC > 0) | (tab_sig$DEA_lab_logFC < 0 & tab_sig$DEA_lit_logFC < 0),]
```


## Merged with significant RNA-Seq (Union)
```{r}
g <- unique(c(dea_lab_sig$Genes, dea_lit_sig$Genes))

tab_sig_union <- Reduce(merge, list(da, dea_lab[dea_lab$Genes %in% g,], dea_lit[dea_lit$Genes %in% g,]))
head(tab_sig_union)

dim(tab_sig_union)
```

## Changed in same direction from both RNA-Seq (significant)
```{r}
tab_dir_sig <- tab_sig[(tab_sig$DEA_lab_logFC > 0 & tab_sig$DEA_lit_logFC > 0) | (tab_sig$DEA_lab_logFC < 0 & tab_sig$DEA_lit_logFC < 0),]
```



# Scatter plots

## ATAC + RNA

### ATAC vs RNA-Seq (lab)
```{r}
p1 <- ggplot(tab, aes(x = DA_logFC, y = DEA_lab_logFC)) +
  geom_point(size = 2, alpha = 0.5) +
  xlab("ATAC logFC") +
  ylab("RNA (lab) logFC")
```

### ATAC vs RNA-Seq (lit)
```{r}
p2 <- ggplot(tab, aes(x = DA_logFC, y = DEA_lit_logFC)) +
  geom_point(size = 2, alpha = 0.5) +
  xlab("ATAC logFC") +
  ylab("RNA (literature) logFC")
```

### RNA-Seq (lab) vs RNA-Seq (lit)
```{r}
p3 <- ggplot(tab, aes(x = DEA_lab_logFC, y = DEA_lit_logFC)) +
  geom_point(size = 2, alpha = 0.5) +
  xlab("RNA (lab) logFC") +
  ylab("RNA (literature) logFC")
```


## ATAC + RNA (significant) (Intersection)

### ATAC vs RNA-Seq (lab)
```{r}
p4 <- ggplot(tab_sig, aes(x = DA_logFC, y = DEA_lab_logFC)) +
  geom_point(size = 2, alpha = 0.5) +
  xlab("ATAC logFC") +
  ylab("RNA (lab) logFC")
```

### ATAC vs RNA-Seq (lit)
```{r}
p5 <- ggplot(tab_sig, aes(x = DA_logFC, y = DEA_lit_logFC)) +
  geom_point(size = 2, alpha = 0.5) +
  xlab("ATAC logFC") +
  ylab("RNA (literature) logFC")
```

### RNA-Seq (lab) vs RNA-Seq (lit)
```{r}
p6 <- ggplot(tab_sig, aes(x = DEA_lab_logFC, y = DEA_lit_logFC)) +
  geom_point(size = 2, alpha = 0.5) +
  xlab("RNA (lab) logFC") +
  ylab("RNA (literature) logFC")
```



## ATAC + RNA (significant) (Union)

### ATAC vs RNA-Seq (lab)
```{r}
p13 <- ggplot(tab_sig_union, aes(x = DA_logFC, y = DEA_lab_logFC)) +
  geom_point(size = 2, alpha = 0.5) +
  xlab("ATAC logFC") +
  ylab("RNA (lab) logFC")
```

### ATAC vs RNA-Seq (lit)
```{r}
p14 <- ggplot(tab_sig_union, aes(x = DA_logFC, y = DEA_lit_logFC)) +
  geom_point(size = 2, alpha = 0.5) +
  xlab("ATAC logFC") +
  ylab("RNA (literature) logFC")
```

### RNA-Seq (lab) vs RNA-Seq (lit)
```{r}
p15 <- ggplot(tab_sig_union, aes(x = DEA_lab_logFC, y = DEA_lit_logFC)) +
  geom_point(size = 2, alpha = 0.5) +
  xlab("RNA (lab) logFC") +
  ylab("RNA (literature) logFC")
```




## ATAC + RNA (changed in same direction)

### ATAC vs RNA-Seq (lab)
```{r}
p7 <- ggplot(tab_dir, aes(x = DA_logFC, y = DEA_lab_logFC)) +
  geom_point(size = 2, alpha = 0.5) +
  xlab("ATAC logFC") +
  ylab("RNA (lab) logFC")
```

### ATAC vs RNA-Seq (lit)
```{r}
p8 <- ggplot(tab_dir, aes(x = DA_logFC, y = DEA_lit_logFC)) +
  geom_point(size = 2, alpha = 0.5) +
  xlab("ATAC logFC") +
  ylab("RNA (literature) logFC")
```

### RNA-Seq (lab) vs RNA-Seq (lit)
```{r}
p9 <- ggplot(tab_dir, aes(x = DEA_lab_logFC, y = DEA_lit_logFC)) +
  geom_point(size = 2, alpha = 0.5) +
  xlab("RNA (lab) logFC") +
  ylab("RNA (literature) logFC")
```


## ATAC + RNA (changed in same direction + significant)

### ATAC vs RNA-Seq (lab)
```{r}
p10 <- ggplot(tab_dir_sig, aes(x = DA_logFC, y = DEA_lab_logFC)) +
  geom_point(size = 2, alpha = 0.5) +
  xlab("ATAC logFC") +
  ylab("RNA (lab) logFC")
```

### ATAC vs RNA-Seq (lit)
```{r}
p11 <- ggplot(tab_dir_sig, aes(x = DA_logFC, y = DEA_lit_logFC)) +
  geom_point(size = 2, alpha = 0.5) +
  xlab("ATAC logFC") +
  ylab("RNA (literature) logFC")
```

### RNA-Seq (lab) vs RNA-Seq (lit)
```{r}
p12 <- ggplot(tab_dir_sig, aes(x = DEA_lab_logFC, y = DEA_lit_logFC)) +
  geom_point(size = 2, alpha = 0.5) +
  xlab("RNA (lab) logFC") +
  ylab("RNA (literature) logFC")
```



# Results

## ATAC + RNA
```{r, fig.width=11, fig.height=4.5, fig.align='center'}
p1 | p2 | p3
```

## ATAC + RNA (significant) (Intersection)
```{r, fig.width=11, fig.height=4.5, fig.align='center'}
p4 | p5 | p6
```

## ATAC + RNA (significant) (Union)
```{r, fig.width=11, fig.height=4.5, fig.align='center'}
p13 | p14 | p15
```


## ATAC + RNA (changed in same direction)
```{r, fig.width=11, fig.height=4.5, fig.align='center'}
p7 | p8 | p9
```

## ATAC + RNA (changed in same direction + significant)
```{r, fig.width=11, fig.height=4.5, fig.align='center'}
p10 | p11 | p12
```


# References
```{r overlap-matrix-rGREAT-43 }
report::cite_packages(session = sessionInfo())
```

# SessionInfo
```{r overlap-matrix-rGREAT-44 }
devtools::session_info() %>%
  details::details()
```
