---
title: "ATAC-Seq: Differential analysis (ID4)"
author: "Deepak Tanwar"
date: "<b>Created on:</b> 2021-01-20 <br> <b>Updated on:</b> `r Sys.Date()`"
output:
  rmdformats::readthedown:
    highlight: tango
    self_contained: true
    thumbnails: true
    lightbox: true
    gallery: false
    keep_md: no
    number_sections: no
    fig_width: 8
    fig_height: 8
    fig_caption: true
    df_print: paged
    code_folding: show
  fontsize: 12pt
  geometry: margin=1in
  documentclass: article
# bibliography: references.bib
link-citations: yes
---

# Library
```{r annotate-id4-peaks-1, message=FALSE, warning=FALSE}
library(GenomicRanges)
```

# Annotating peaks
```{r annotate-id4-peaks-2 }
anno_peaks <- function(peak_file) {
  peak <- GRanges(read.table(peak_file,
    header = F,
    col.names = c(
      "Chr", "Start", "End", "Name", "Score", "Strand",
      "FoldChange", "negLog10pvalue", "negLog10qvalue",
      "Summit"
    ), stringsAsFactors = F
  ))
  pa <- data.frame(
    plgINS:::annoPeaks(
      peaks = peak,
      gtf = "/mnt/IM/reference/annotation/gencodeM18/gencode.vM18.chr_patch_hapl_scaff.annotation.gtf.gz",
      proximal = 1000
    )
  )
  colnames(pa) <- gsub(pattern = "_pa|\\.", replacement = "", x = colnames(pa))
  return(pa)
}

dim <- anno_peaks(peak_file = "input/Dim_NFF_peaks.narrowPeak")
bright <- anno_peaks(peak_file = "input/Bright_NFF_peaks.narrowPeak")
dim_bright <- anno_peaks(peak_file = "input/Dim_Bright_NFF_peaks.narrowPeak")
save(dim, bright, dim_bright, file = "./output/id4_annotated_peaks.RData")
```

# SessionInfo
```{r annotate-id4-peaks-3 }
devtools::session_info()
```
