---
title: "Figure4"
author: "Deepak Tanwar"
date: "<b>Created on:</b> 2020-07-30 <br> <b>Updated on:</b> `r Sys.Date()`"
output:
  rmdformats::readthedown:
    highlight: tango
    self_contained: true
    thumbnails: true
    lightbox: true
    gallery: false
    keep_md: no
    number_sections: no
    fig_width: 8
    fig_height: 8
    fig_caption: true
    df_print: paged
    code_folding: hide
  fontsize: 12pt
  geometry: margin=1in
  documentclass: article
# bibliography: references.bib
link-citations: yes
---


# Library
```{r overlap-matrix-rGREAT-1, warning = F, message = F}
library(tidyverse)
library(report)
library(ggpubr)
library(patchwork)
library(viridis)
library(ggfortify)
library(pheatmap)
library(ComplexHeatmap)
library(circlize)
library(readxl)
library(plyr)
library(reshape2)
library(extrafont)
library(ggthemes)
library(EnhancedVolcano)
library(SummarizedExperiment)
library(csaw)
library(edgeR)
library(seriation)
library(ggsci)
ht_opt$message <- FALSE
```


# Data
```{r}
load("input/repeat_elements_atac_diff_chromatin_accessibility.RData")
tab <- data.frame(rowData(data), check.names = F, stringsAsFactors = F)


load("input/data_pData.RData")
load("input/gencode.vM18.anno.RData")

ge <- data.frame(symbol = rownames(data$cpm), data$cpm)
ge_an <- merge(anno[,c(3,5)], ge)
ge_an <- ge_an[!duplicated(ge_an),]
re_ele <- subset(ge_an, type %in% c("LTR", "LINE", "SINE"))
re_ele <- re_ele[,c(grep(pattern = "symbol|type", x = colnames(re_ele)),
                    grep(pattern = "PND15", x = colnames(re_ele)),
                    grep(pattern = "Adult", x = colnames(re_ele)))]
rownames(re_ele) <- re_ele$symbol

pData <- data$pData
```

# Colors
```{r}
cols <- magma(n = 10)[c(2, 5, 8)]
cols <- c(PND8 = cols[3], PND15 = cols[2], Adult = cols[1])
```

# Figures
## A
```{r, fig.align='center', fig.height=6, fig.width=8}
th <- theme_bw(base_size = 24) +

  theme(
    legend.background = element_rect(),

    # title, subtitle, and caption
    plot.title = element_text(
      angle = 0,
      size = 18,
      face = "bold",
      vjust = 1
    ),
    plot.subtitle = element_text(
      angle = 0,
      size = 14,
      face = "plain",
      vjust = 1
    ),
    plot.caption = element_text(
      angle = 0,
      size = 14,
      face = "plain",
      vjust = 1
    ),

    # axis text
    axis.text.x = element_text(
      angle = 0,
      size = 16,
      vjust = 1
    ),
    axis.text.y = element_text(
      angle = 0,
      size = 16,
      vjust = 1
    ),
    axis.title = element_text(
      size = 18
    ),

    # legend
    legend.position = "top",
    legend.key = element_blank(),
    legend.key.size = unit(0.5, "cm"),
    legend.text = element_text(
      size = 16
    ),
    title = element_text(
      size = 14
    ),
    legend.title = element_text(size = 18, face = "bold")
  )

legendLabels <- c(
  paste("Not Significant"),
  expression(adjusted ~ italic(P) - 0.05 ~ "&" ~ log[2] ~ -1),
  expression(adjusted ~ italic(P) - 0.05 ~ "&" ~ log[2] ~ 1)
)
```

```{r}
tab$Significant <- "Not Significant"
tab$Significant[tab$`diffEnrichment-qvalue` <= 0.05 & abs(tab$`diffEnrichment-logFC`) >= 1] <- "Negative"
# tab$Significant[tab$`diffAccessibility-qvalue` <= 0.05 & tab$`diffAccessibility-logFC` >= 1] <- "Positive"

tab$Significant[tab$Significant == "Negative"] <- ifelse(tab$`diffEnrichment-logFC`[tab$Significant == "Negative"] > 0, "Positive", "Negative")

tab$Significant[tab$Significant == "Negative"] <- "Low acc. in Adult"
tab$Significant[tab$Significant == "Positive"] <- "High acc. in Adult"

tab$Significant <- factor(x = tab$Significant, levels = unique(tab$Significant))

a <- ggplot(subset(tab, class_id %in% c("LTR", "LINE", "SINE")), aes(x = `diffEnrichment-logFC`, y = -log10(`diffEnrichment-qvalue`))) +
  geom_point(aes(color = Significant), size = 2, alpha = 0.5) +
  scale_color_manual(values = c("black", "#BC3C29FF", "#0072B5FF")) +
  xlim(
    -(ceiling(max(abs(tab$`diffEnrichment-logFC`)))),
    ceiling(max(abs(tab$`diffEnrichment-logFC`)))
  ) +
  ylim(0, max(-log10(tab$`diffEnrichment-qvalue`), na.rm = TRUE) + 1) +
  xlab(bquote(~ Log[2] ~ "fold change (FC)")) +
  ylab(bquote(~ -Log[10] ~ adjusted ~ italic(P))) +
  labs(
    title = "",
    subtitle = "", caption = paste0("Total = ", nrow(tab), " variables")
  ) +
  th +
  geom_vline(
    xintercept = c(-1, 1),
    linetype = "longdash",
    colour = "black",
    size = 0.4
  ) +
  geom_hline(
    yintercept = -log10(0.05),
    linetype = "longdash",
    colour = "black",
    size = 0.4
  ) +
  guides(
    colour = guide_legend(
      order = 1,
      override.aes = list(
        size = 5
      )
    )
  ) +
  scale_shape_manual(
    values = c(
      `Not Significant` = 19,
      Positive = 19,
      Negative = 19
    ),
    labels = c(
      `Not Significant` = legendLabels[1],
      Positive = legendLabels[3],
      Negative = legendLabels[2]
    )
  ) +
  facet_wrap(~class_id) +
  theme(legend.justification="bottom",
          legend.margin=margin(0,0,0,0),
          legend.box.margin=margin(0,0,-20,0),
          plot.margin=grid::unit(c(1,1,1,1), "mm"))

an <- split(x = tab, f = tab$class_id)
an1 <- lapply(an, function(x){
  return(data.frame(x = c(-5, 5), 
                    y = c(5,5), 
                    label = c(paste("n =", 
                                    nrow(x[x$`diffEnrichment-logFC` <= 1 & x$`diffEnrichment-qvalue` <= 0.05,])),   
                              paste("n =", 
                                    nrow(x[x$`diffEnrichment-logFC` >= 1 & x$`diffEnrichment-qvalue` <= 0.05,]))),
                    col = c("#BC3C29FF", "#0072B5FF")))
})

anno <- ldply(an1, data.frame, .id = "class_id")
anno <- subset(anno, class_id %in% c("LTR", "LINE", "SINE"))

a + geom_label(
    data = anno, aes(x = x, y = y, label = label),
    color = rep(c("#BC3C29FF", "#0072B5FF"),3),
    size = 5, angle = 45, fontface = "bold"
  )
```

## B
```{r}
re_ele <- re_ele[grep(pattern = "\\?", x = re_ele$symbol, value = T, invert = T),]
mat <- re_ele[,3:ncol(re_ele)]
mat <- data.matrix(mat - rowMeans(mat[, grep(pattern = "PND15", x = colnames(mat), value = T)]))

o <- seriate(max(mat) - mat)

mat <- mat[get_order(o), ]

rd <- re_ele[rownames(mat),1:2]

cd <- pData[colnames(mat),]

cols <- magma(n = 10)[c(2, 5, 8)]
cols <- c(PND8 = cols[3], PND15 = cols[2], Adult = cols[1])

ele_cols <- RColorBrewer::brewer.pal(n = 3, name = "Dark2")
# names(kmeans_cols) <- letters[1:5]
names(ele_cols) <- c("LINE", "LTR", "SINE")



# col_fun <- colorRamp2(c(-2, 0, 2), viridis(n = 3))
b <- Heatmap(
    matrix = mat,
    col = viridis(n = 100),
    cluster_rows = F,
    # row_order = get_order(o, 1),
    # column_order = get_order(o, 2),
    cluster_columns = T,
    clustering_method_columns = "ward.D2",
    show_row_names = T,
    show_column_names = F,
    name = "Enrichment (log)",
    row_split = rd$type,
    row_title = NULL,
    use_raster = FALSE,
    top_annotation = HeatmapAnnotation(
      df = cd[, 1, drop = FALSE],
      col = list(Group = cols[2:3]),
    simple_anno_size = unit(5, "mm"), 
    show_annotation_name = TRUE, 
    annotation_name_side = "right", 
    annotation_legend_param = gpar(fontsize = 3), 
    annotation_name_rot = 0, 
    annotation_name_gp = gpar(fontsize = 10)
    ),
      left_annotation = HeatmapAnnotation(
    Category = rd$type, which = "row",
    col = list(Category = ele_cols),
    simple_anno_size = unit(5, "mm"), 
    show_annotation_name = TRUE, 
    annotation_name_side = "bottom", 
    annotation_legend_param = gpar(fontsize = 3), 
    annotation_name_rot = 0, 
    annotation_name_gp = gpar(fontsize = 10)
  ), height = unit(x = 4, units = "in"), width = unit(x = 3, units = "in"),
  heatmap_legend_param = list(
    title = expression(bold("Gene expression (" ~ log[2] ~ "CPM)")),
    legend_height = unit(5, "cm"),
    title_position = "leftcenter-rot"
)
)

b

png(filename = "output/b.png", width = 7, height = 7, units = "in", res = 320)
draw(object = b, 
     heatmap_legend_side = "right", 
     annotation_legend_side = "right",
     merge_legends = TRUE,
     align_annotation_legend = "")

dev.off()

pdf(file = "output/b.pdf", width = 7, height = 7)
draw(object = b, 
     heatmap_legend_side = "right", 
     annotation_legend_side = "right",
     merge_legends = TRUE,
     align_annotation_legend = "")

dev.off()

```


<!-- ## C -->
<!-- ```{r, eval=FALSE} -->
<!-- library(parallel) -->
<!-- library(rtracklayer) -->

<!-- ltr <- GRanges(subset(tab, class_id %in% "LTR")) -->
<!-- line <- GRanges(subset(tab, class_id %in% "LINE")) -->
<!-- sine <- GRanges(subset(tab, class_id %in% "SINE")) -->

<!-- ll <- list(LTR = ltr, -->
<!--            LINE = line, -->
<!--            SINE = sine) -->

<!-- cov_adult <- mclapply(ll, function(x) import(con = "input/Adult_NF.bw", selection = x),  -->
<!--                       mc.preschedule = F, mc.cores = length(ll)) -->

<!-- cov_pnd15 <- mclapply(ll, function(x) import(con = "input/PND15_NF.bw", selection = x),  -->
<!--                       mc.preschedule = F, mc.cores = length(ll)) -->

<!-- library(ChIPseeker) -->
<!-- library(TxDb.Mmusculus.UCSC.mm10.knownGene) -->
<!-- library(gUtils) -->


<!-- get_tag_mat <- function(peak, cov){ -->
<!--   pr <- promoters(x = gUtils::gr.mid(peak), upstream = 150001, downstream = 15000) -->
<!--   tagMatrix <- getTagMatrix(peak = cov, windows=pr) -->
<!--   return(tagMatrix) -->
<!-- } -->

<!-- tag_adult <- list() -->
<!--   for(i in 1:3){ -->
<!--   n <- names(ll)[i] -->
<!--   peak <- ll[[n]] -->
<!--   cov <- cov_adult[[n]] -->
<!--   tag_adult[[n]] <- get_tag_mat(peak = peak, cov = cov) -->
<!--   invisible(gc()) -->
<!-- } -->

<!-- tag_pnd15 <- list() -->
<!--   foreach(i in 1:3){ -->
<!--   n <- names(ll)[i] -->
<!--   peak <- ll[[n]] -->
<!--   cov <- cov_pnd15[[n]] -->
<!--   tag_pnd15[[n]] <- get_tag_mat(peak = peak, cov = cov) -->
<!--   invisible(gc()) -->
<!-- } -->

<!-- save(tag_adult, tag_pnd15, file = "output/tags_mat.RData") -->

<!-- plotAvgProf(tagMatrix = tagMatrix,xlim = c(-4999,5000), -->
<!--             xlab="", ylab = "") -->
<!-- ``` -->

## C
```{r}
tf_files <- list.files(path = "input", pattern = "ive_homer", full.names = T)
names(tf_files) <- gsub(pattern = "input/|_homer.txt", replacement = "", x = tf_files)

tf <- lapply(tf_files, function(x) {
  a <- read.delim(file = x, header = T, check.names = F)
  a <- a[a[,5] <= 0.05,]
  return(a)
})

tf_df <- ldply(tf, data.frame)
colnames(tf_df)[1] <- "Group"
tf_df$Group <- gsub(pattern = "_", replacement = " ", x = tf_df$Group)

tf_df$Genes <- str_to_title(tolower(gsub(pattern = "_.*", replacement = "", x = tf_df$Motif.Name)))
tf_df$Motif <- gsub(pattern = "_.*", replacement = "", x = tf_df$Motif.Name)

tf_df$Group[tf_df$Group == "Adult Positive"] <- "High acc. in Adult"
tf_df$Group[tf_df$Group == "Adult Negative"] <- "Low acc. in Adult"


load("input/limma_SC_Controls.RData")
dea <- dea.limma$`PND15 vs Adult`
tf_dea <- merge(tf_df, dea)

c <- ggplot(tf_dea, aes(x = Motif, y = Group)) +
  geom_point(aes(size = -log10(adj.P.Val), color = logFC)) +
    scale_colour_gradient2(low = "#BC3C29FF", mid = "grey", high = "#0072B5FF", limits=c(-5,5), breaks = c(-5,0,5)) +
  # scale_colour_gradientn(colors = c("#BC3C29FF","grey", "#0072B5FF"), guide=guide_colorbar(reverse=FALSE))+
  xlab("") +
  ylab("") +
  th +
  theme(
    axis.text.x = element_text(angle = 90, size = 10),
    axis.text.y = element_text(size = 12)
  ) +
  labs(color = expression(~ Log[2] ~ fold-change), size = expression(adjusted ~ italic(P)))
c

```


# References
```{r overlap-matrix-rGREAT-43 }
report::cite_packages(session = sessionInfo())
```

# SessionInfo
```{r overlap-matrix-rGREAT-44 }
devtools::session_info() %>%
  details::details()
```