---
title: "RNA-Seq: DEU using DEXSeq"
author: "Deepak Tanwar"
date: "<b>Created on:</b> 2019-08-07 <br> <b>Updated on:</b> `r Sys.Date()`"
output:
  rmdformats::readthedown:
    highlight: tango
    self_contained: true
    thumbnails: true
    lightbox: true
    gallery: false
    keep_md: no
    number_sections: no
    fig_width: 8
    fig_height: 8
    fig_caption: true
    df_print: paged
    code_folding: show
  fontsize: 12pt
  geometry: margin=1in
  documentclass: article
# bibliography: references.bib
link-citations: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Libraries required

```{r, message=FALSE, warning=FALSE}
library(DEXSeq)
library(dplyr)
library(stringr)
library(tidyverse)
```

# FC list

```{r}
files <- list.files(path = "./input/featureCounts", pattern = ".RData", full.names = T)

featureCounts <- lapply(files, function(x) {
  load(x)
  res <- data.frame(fc$annotation, fc$counts, stringsAsFactors = F)
  res %>%
    mutate_all(trimws) -> res
  return(res)
})

fc <- NULL

for (i in 1:length(featureCounts)) {
  if (i == 1) {
    fc <- featureCounts[[i]]
  } else {
    fc <- data.frame(fc, featureCounts[[i]][, 7, drop = F])
  }
}

colnames(fc) <- gsub(pattern = ".bam", replacement = "", x = colnames(fc))
colnames(fc) <- gsub(pattern = "\\.", replacement = "_", x = colnames(fc))

# Mapping Ensemble IDs to GeneSymbols
load("./input/gencode.vM18.anno.RData")

anno %>%
  mutate_all(trimws) -> anno

anno <- anno[, c(2, 3, 7)]
rownames(anno) <- NULL
anno <- anno[!duplicated(anno$ensembl), ]

rownames(anno) <- anno$ensembl

fc <- data.frame(anno[fc$GeneID, ], fc, stringsAsFactors = F)

df <- fc[1:1000, ]

fc$gID <- fc$symbol
fc$gID[is.na(fc$gID)] <- fc$symbol1[is.na(fc$gID)]
fc$gID[is.na(fc$gID)] <- fc$GeneID[is.na(fc$gID)]

fc$GeneID <- fc$gID
fc <- fc[, -c(1:3, ncol(fc))]

# Import pheno .tsv file
pheno <- read.table(file = "./input/dataset_pnd_adult.tsv", sep = "	", header = TRUE, stringsAsFactors = F)

rownames(pheno) <- pheno$Samples

# Now, make a list with the featureCounts table and the pheno data from the .tsv file
phenoFClist <- list(featureCounts = fc, pheno = pheno[colnames(fc)[7:ncol(fc)], ])

save(phenoFClist,
  file = paste0("./output/featureCounts_pheno_list.RData"),
  compress = T,
  compression_level = 3
)
```

# DEXSeq analysis


## Function
```{r}
dexseq_analysis <- function(counts, pheno, design) {
  countFile <- counts
  sampleData <- pheno

  countFile <- countFile[order(countFile$GeneID, countFile$Start), ]
  countFile$exon_id <- paste(countFile$GeneID,
    countFile$Chr,
    countFile$Start,
    countFile$End,
    countFile$Strand,
    sep = "_"
  )

  countFile2 <- countFile[!duplicated(countFile$exon_id), ]
  countFile2 <- countFile2[, -ncol(countFile2)]

  w <- c(which(!duplicated(countFile2$GeneID)), nrow(countFile2) + 1)
  exonid <- paste0("E", unlist(lapply(w[-1] - w[-length(w)], FUN = function(x) 1:x)))
  featureID <- exonid

  dxd <- DEXSeqDataSet(
    countData = data.matrix(countFile2[, -1:-6]),
    sampleData = sampleData,
    design = design,
    featureID = featureID,
    groupID = countFile2$GeneID,
    featureRanges = GRanges(countFile2$Chr,
      IRanges(
        start = as.numeric(countFile2$Start),
        end = as.numeric(countFile2$End)
      ),
      strand = countFile2$Strand
    )
  )

  dxd <- estimateSizeFactors(dxd)
  dxd <- estimateDispersions(dxd, BPPARAM = BiocParallel::MulticoreParam(workers = 16))

  # remove "TESTVAR:exon" term from formula, set this as reducedModel
  redMod <- terms(design)
  f <- grep(pattern = ":exon", x = attr(redMod, "term.labels"))
  redMod <- drop.terms(redMod, dropx = f)
  redMod <- reformulate(attr(redMod, "term.labels"))

  # test for DEU
  dxd <- testForDEU(
    object = dxd,
    fullModel = design(dxd),
    reducedModel = redMod,
    BPPARAM = BiocParallel::MulticoreParam(workers = 16)
  )

  # estimate exon fc
  dxd <- estimateExonFoldChanges(
    object = dxd,
    fitExpToVar = "Group",
    BPPARAM = BiocParallel::MulticoreParam(workers = 16),
    denominator = , maxRowsMF =
    )

  dxr <- DEXSeqResults(dxd)
}
```

## PND8 vs PND15
```{r}
pheno1 <- phenoFClist$pheno[phenoFClist$pheno$Group == "PND8" | phenoFClist$pheno$Group == "PND15", ]
pheno1$Group <- as.factor(pheno1$Group)
pheno1$Group <- relevel(pheno1$Group, ref = "PND8")


fc1 <- data.frame(phenoFClist$featureCounts[, 1:6],
  phenoFClist$featureCounts[, rownames(pheno1)],
  stringsAsFactors = F
)

# design formula
design <- as.formula(paste0("~", "Group", " + ", "exon", " + ", "Group", ":exon"))

dexseq.pnd8.pnd15 <- dexseq_analysis(counts = fc1, pheno = pheno1, design = design)

# save dxr and design matrix in a list object
dexseqList <- list(dxr = dexseq.pnd8.pnd15, design_matrix = design)
save(dexseqList, file = "./output/PND8vsPND15.RData")

# Create DEXSeq HTML report file
DEXSeqHTML(
  object = dexseq.pnd8.pnd15,
  FDR = 0.1,
  color = c("#FF000080", "#0000FF80"),
  BPPARAM = BiocParallel::MulticoreParam(workers = 16),
  path = "./output/PND8vsPND15/",
  file = "PND8vsPND15.html",
  fitExpToVar = "Group"
)
```


## PND15 vs Adult
```{r}
pheno2 <- phenoFClist$pheno[phenoFClist$pheno$Group == "Adult" | phenoFClist$pheno$Group == "PND15", ]
pheno2$Group <- as.factor(pheno2$Group)
pheno2$Group <- relevel(pheno2$Group, ref = "PND15")

fc2 <- data.frame(phenoFClist$featureCounts[, 1:6],
  phenoFClist$featureCounts[, rownames(pheno2)],
  stringsAsFactors = F
)

# design formula
design <- as.formula(paste0("~", "Group", " + ", "exon", " + ", "Group", ":exon"))

dexseq.pnd15.Adult <- dexseq_analysis(counts = fc2, pheno = pheno2, design = design)

# save dxr and design matrix in a list object
dexseqList <- list(dxr = dexseq.pnd15.Adult, design_matrix = design)
save(dexseqList, file = "./output/PND15vsAdult.RData")

# Create DEXSeq HTML report file
DEXSeqHTML(
  object = dexseq.pnd15.Adult,
  FDR = 0.1,
  color = c("#FF000080", "#0000FF80"),
  BPPARAM = BiocParallel::MulticoreParam(workers = 16),
  path = "./output/PND15vsAdult/",
  file = "PND15vsAdult.html",
  fitExpToVar = "Group"
)
```

# SessionInfo
```{r}
devtools::session_info()
```