---
title: "RNA-Seq: differential expression analysis of controls (lab + literature)"
author: "Deepak Tanwar"
date: "<b>Created on:</b> 2020-09-11 <br> <b>Updated on:</b> `r Sys.Date()`"
output:
  rmdformats::readthedown:
    highlight: tango
    self_contained: true
    thumbnails: true
    lightbox: true
    gallery: false
    keep_md: no
    number_sections: no
    fig_width: 8
    fig_height: 8
    fig_caption: true
    df_print: paged
    code_folding: show
  fontsize: 12pt
  geometry: margin=1in
  documentclass: article
# bibliography: references.bib
link-citations: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Libraries required

```{r, message=FALSE, warning=FALSE}
library(limma)
library(edgeR)
library(plgINS)
library(sva)
library(ggplot2)
library(dplyr)
library(patchwork)
library(SummarizedExperiment)
library(reshape2)
library(autoplotly)
library(pheatmap)
library(viridis)
library(RColorBrewer)
library(ggplotify)
```

# Load salmon object
```{r, message=FALSE, warning=FALSE}
load("input/SC_controls_rnaseq_lab_lit_salmon.tds.RData")
salmon <- salmon[c(11:18, 1:10, 19)]
```

# Differeitial analysis using `limma`

## PND8 vs PND15
```{r }
design <- model.matrix(~ 0 + Group, data = salmon@phenoData[grep(pattern = "PND8|PND15", x = salmon@phenoData$Group), ])
colnames(design) <- gsub(pattern = "Group", replacement = "", x = colnames(design))

y <- DGEList(counts = salmon@gene.counts[, grep(pattern = "PND8|PND15", x = colnames(salmon@gene.counts))])
keep <- filterByExpr(y, design, min.count = 15)
y <- y[keep, ]

dds <- calcNormFactors(y)
v <- voom(dds, design = design)

contrast.matrix <- makeContrasts(PND15 - PND8, levels = design)
fit <- lmFit(v)

fit2 <- contrasts.fit(fit, contrast.matrix)
fit2 <- eBayes(fit2)

pnd8.pnd15 <- as.data.frame(topTable(fit2, coef = 1, number = Inf))
pnd8.pnd15 <- data.frame(Genes = rownames(pnd8.pnd15), pnd8.pnd15, stringsAsFactors = F)
pnd8.pnd15_sig <- pnd8.pnd15[abs(pnd8.pnd15$logFC) >= 1 & pnd8.pnd15$adj.P.Val <= 0.05, ]
```

<!-- ## PND15 vs PNW8 -->
<!-- ```{r } -->
<!-- design <- model.matrix(~ 0 + Group, -->
<!--   data = salmon@phenoData[grep(pattern = "PND15|PNW8", x = salmon@phenoData$Group), ] -->
<!-- ) -->
<!-- colnames(design) <- gsub(pattern = "Group", replacement = "", x = colnames(design)) -->

<!-- y <- DGEList(counts = salmon@gene.counts[, grep(pattern = "PND15|PNW8", x = colnames(salmon@gene.counts))]) -->
<!-- keep <- filterByExpr(y, design, min.count = 15) -->
<!-- y <- y[keep, ] -->

<!-- dds <- calcNormFactors(y) -->
<!-- v <- voom(dds, design = design) -->

<!-- contrast.matrix <- makeContrasts(PNW8 - PND15, levels = design) -->
<!-- fit <- lmFit(v) -->

<!-- fit2 <- contrasts.fit(fit, contrast.matrix) -->
<!-- fit2 <- eBayes(fit2) -->


<!-- pnd15_o.pnw8 <- as.data.frame(topTable(fit2, coef = 1, number = Inf)) -->
<!-- pnd15_o.pnw8 <- data.frame(Genes = rownames(pnd15_o.pnw8), pnd15_o.pnw8, stringsAsFactors = F) -->
<!-- pnd15_o.pnw8_sig <- pnd15_o.pnw8[abs(pnd15_o.pnw8$logFC) >= 1 & pnd15_o.pnw8$adj.P.Val <= 0.05, ] -->
<!-- ``` -->

## PND14 vs PNW8
```{r }
design <- model.matrix(~ 0 + Group,
  data = salmon@phenoData[grep(pattern = "PND14|PNW8", x = salmon@phenoData$Group), ]
)
colnames(design) <- gsub(pattern = "Group", replacement = "", x = colnames(design))

y <- DGEList(counts = salmon@gene.counts[, grep(pattern = "PND14|PNW8", x = colnames(salmon@gene.counts))])
keep <- filterByExpr(y, design, min.count = 15)
y <- y[keep, ]

dds <- calcNormFactors(y)
v <- voom(dds, design = design)

contrast.matrix <- makeContrasts(PNW8 - PND14, levels = design)

# fit <- lmFit(v)
# 
# fit2 <- contrasts.fit(fit, contrast.matrix)
# try(
#   fit2 <- eBayes(fit2)
# )
# try(
#   pnd14.pnw8 <- as.data.frame(topTable(fit2, coef = 1, number = Inf))
# )
# try(
#   pnd14.pnw8 <- data.frame(Genes = rownames(pnd14.pnw8), pnd14.pnw8, stringsAsFactors = F)
# )
# try(
#   pnd14.pnw8_sig <- pnd14.pnw8[abs(pnd14.pnw8$logFC) >= 1 & pnd14.pnw8$adj.P.Val <= 0.05, ]
# )

pnd14.pnw8_lfc <- v$E[,2] - v$E[,1]
pnd14.pnw8 <- data.frame(Genes = names(pnd14.pnw8_lfc), logFC = pnd14.pnw8_lfc, AveExpr = rowMeans(v$E), 
                         t = NA, P.Value = NA, adj.P.Val = NA, B = NA)
```


<!-- ## PND14 + PND15 vs PNW8 -->
<!-- ```{r } -->
<!-- design <- model.matrix(~ 0 + Group1 + Batch, -->
<!--   data = salmon@phenoData[grep(pattern = "PND14|PND15|PNW8", x = salmon@phenoData$Group), ] -->
<!-- ) -->
<!-- colnames(design) <- gsub(pattern = "Group1", replacement = "", x = colnames(design)) -->

<!-- y <- DGEList(counts = salmon@gene.counts[, grep(pattern = "PND14|PND15|PNW8", x = colnames(salmon@gene.counts))]) -->
<!-- keep <- filterByExpr(y, design, min.count = 15) -->
<!-- y <- y[keep, ] -->

<!-- dds <- calcNormFactors(y) -->
<!-- v <- voom(dds, design = design) -->

<!-- contrast.matrix <- makeContrasts(PNW8 - PND14_PND15, levels = design) -->
<!-- fit <- lmFit(v) -->

<!-- fit2 <- contrasts.fit(fit, contrast.matrix) -->
<!-- fit2 <- eBayes(fit2) -->


<!-- pnd15.pnw8 <- as.data.frame(topTable(fit2, coef = 1, number = Inf)) -->
<!-- pnd15.pnw8 <- data.frame(Genes = rownames(pnd15.pnw8), pnd15.pnw8, stringsAsFactors = F) -->
<!-- pnd15.pnw8_sig <- pnd15.pnw8[abs(pnd15.pnw8$logFC) >= 1 & pnd15.pnw8$adj.P.Val <= 0.05, ] -->
<!-- ``` -->


# `cpm` Counts data and pData

```{r }
y <- DGEList(counts = salmon@gene.counts)
dds <- calcNormFactors(y)
cpm <- cpm(dds, log = T)

data <- list(cpm = cpm, pData = salmon@phenoData)

save(data,
  file = "./output/data_pData.RData", compress = T,
  compression_level = 3
)
```

# Results
```{r }
dea.list <- list(
  `pnd8 vs pnd15` = as.DEA(pnd8.pnd15),
  `pnd14 vs pnw8` = as.DEA(pnd14.pnw8)
)

# normCounts <- v$E
# 
# voomEList <- v

dea.limma <- list(
  `pnd8 vs pnd15` = pnd8.pnd15,
  `pnd14 vs pnw8` = pnd14.pnw8
)
```


## Save RData files
```{r }
save(dea.list,
  file = "./output/dea_SC_Controls_lab_lit.DEA.RData", compress = T,
  compression_level = 3
)

# save(normCounts,
#   file = "./output/normCounts_voom_SC_Controls_lab_lit.RData", compress = T,
#   compression_level = 3
# )

# save(voomEList,
#   file = "./output/voom_EList_SC_Controls_lab_lit.RData", compress = T,
#   compression_level = 3
# )

save(dea.limma,
  file = "./output/limma_SC_Controls_lab_lit.RData", compress = T,
  compression_level = 3
)
```


# MA plots
## PND8 vs PND15
```{r }
res_8_15 <- pnd8.pnd15
res_8_15$threshold <- as.factor(res_8_15$adj.P.Val < 0.05)

p1 <- ggplot(data = res_8_15, aes(
  x = res_8_15$AveExpr,
  y = res_8_15$logFC,
  colour = threshold
)) +
  geom_point(alpha = 0.5, size = 1.8) +
  geom_hline(aes(yintercept = 0), colour = "blue", size = 1) +
  ylim(c(
    -ceiling(max(abs(res_8_15$logFC))),
    ceiling(max(abs(res_8_15$logFC)))
  )) +
  ggtitle("PND8 vs PND15") +
  labs(subtitle = "loess fit") +
  xlab("Mean expression") +
  ylab("Log2 Fold Change") +
  theme(
    plot.title = element_text(face = "bold", size = 20, hjust = 0.5),
    plot.subtitle = element_text(size = 16, hjust = 0.5, face = "italic", color = "blue"),
    axis.title.x = element_text(face = "bold", size = 15),
    axis.text.x = element_text(face = "bold", size = 12),
    legend.title = element_text(face = "bold", size = 15),
    legend.text = element_text(size = 14)
  ) +
  scale_colour_discrete(name = "p.adjusted < 0.05") +
  stat_smooth(se = FALSE, method = "loess", color = "red", formula = y ~ x, size = 1)
```

<!-- ## PND15 vs PNW8 -->
<!-- ```{r } -->
<!-- res_15_w8 <- pnd15.pnw8 -->
<!-- res_15_w8$threshold <- as.factor(res_15_w8$adj.P.Val < 0.05) -->

<!-- p2 <- ggplot(data = res_15_w8, aes( -->
<!--   x = res_15_w8$AveExpr, -->
<!--   y = res_15_w8$logFC, -->
<!--   colour = threshold -->
<!-- )) + -->
<!--   geom_point(alpha = 0.5, size = 1.8) + -->
<!--   geom_hline(aes(yintercept = 0), colour = "blue", size = 1) + -->
<!--   ylim(c( -->
<!--     -ceiling(max(abs(res_15_w8$logFC))), -->
<!--     ceiling(max(abs(res_15_w8$logFC))) -->
<!--   )) + -->
<!--   ggtitle("PND15 vs PNW8") + -->
<!--   labs(subtitle = "loess fit") + -->
<!--   xlab("Mean expression") + -->
<!--   ylab("Log2 Fold Change") + -->
<!--   theme( -->
<!--     plot.title = element_text(face = "bold", size = 20, hjust = 0.5), -->
<!--     plot.subtitle = element_text(size = 16, hjust = 0.5, face = "italic", color = "blue"), -->
<!--     axis.title.x = element_text(face = "bold", size = 15), -->
<!--     axis.text.x = element_text(face = "bold", size = 12), -->
<!--     legend.title = element_text(face = "bold", size = 15), -->
<!--     legend.text = element_text(size = 14) -->
<!--   ) + -->
<!--   scale_colour_discrete(name = "p.adjusted < 0.05") + -->
<!--   stat_smooth(se = FALSE, method = "loess", color = "red", formula = y ~ x, size = 1) -->
<!-- ``` -->

```{r, fig.height=5, fig.width=11}
p1
```


# Heatmap of most variable genes
```{r}
xvar <- apply((data$cpm + 1), 1, var)
genes500 <- head(sort(xvar, decreasing = TRUE), n = 500)
x500 <- data$cpm[rownames(data$cpm) %in% names(genes500), ]

pheatmap(x500, scale = "row", show_rownames = F, col = viridis(100), show_colnames = F, annotation_col = data$pData[,c(1,4)], main = "Clustering of samples based on top 500 variable genes")
```

# PCA
```{r, fig.height=8.5, fig.width=11}
pca <- prcomp(t(data$cpm))
# ggplotly(
  autoplot(pca,
  data = data$pData, colour = "Group",
  frame = TRUE, frame.type = "norm", size = 2
) +
  ggtitle("") + 
    theme(axis.text.x = element_text(size = 20),
        axis.text.y = element_text(size = 20),  
        axis.title.x = element_text(size = 20),
        axis.title.y = element_text(size = 20),
        plot.title = element_text(size = 25, face = "bold"),
        legend.text=element_text(size=20),
        legend.title=element_text(size=20, face = "bold"))
```




# References
```{r }
report::cite_packages(session = sessionInfo())
```


# SessionInfo
```{r }
devtools::session_info() %>%
  details::details()
```
