---
title: "Figure5"
author: "Deepak Tanwar"
date: "<b>Created on:</b> 2020-10-14 <br> <b>Updated on:</b> `r Sys.Date()`"
output:
  rmdformats::readthedown:
    highlight: tango
    self_contained: true
    thumbnails: true
    lightbox: true
    gallery: false
    keep_md: no
    number_sections: no
    fig_width: 8
    fig_height: 8
    fig_caption: true
    df_print: paged
    code_folding: hide
  fontsize: 12pt
  geometry: margin=1in
  documentclass: article
# bibliography: references.bib
link-citations: yes
---


# Libraries and functions
```{r}
source("../bin/functions.R")
```


# A
```{r}
load("input/repeat_elements_atac_diff_chromatin_accessibility_gene_level.RData")

da <- data.frame(rowData(data_ge), stringsAsFactors = F, check.names = F)
atac_da <- da[da$subtype_qvalue <= 0.05 & da$subtype_logFC <= -0.5, ]
atac_da <- atac_da[atac_da$class_id %in% c("LINE", "LTR", "SINE"), ]
r <- rownames(atac_da)

y_atac <- DGEList(assay(data_ge))
colnames(y_atac) <- gsub(pattern = "input/|\\.bam", replacement = "", x = colnames(assay(data_ge)))
colnames(y_atac) <- gsub(pattern = "\\.", replacement = "_", x = colnames(y_atac))
y_atac <- calcNormFactors(y_atac)

atac_cpm <- cpm(y_atac, log = TRUE)
atac_cpm <- atac_cpm[r, ]
atac_cpm <- atac_cpm[, c(grep(pattern = "PND15", x = colnames(atac_cpm)), grep(pattern = "Adult", x = colnames(atac_cpm)))]
```


# B
```{r}
load("input/limma_SC_Controls_lab_lit_tx.RData")

atac_da_s <- da[da$subtype_qvalue <= 0.05 & abs(da$subtype_logFC) >= 0.5, ]
ra <- merge(x = atac_da_s[, 1, drop = FALSE], y = dea.limma$`pnd14 vs pnw8`, all.x = T, by = "row.names")
rownames(ra) <- ra$Row.names

atac_da_s <- atac_da_s[atac_da_s$subtype_logFC > 0, ]
atac_da_s <- atac_da_s[atac_da_s$class_id %in% c("LINE", "LTR", "SINE"), ]
r_s <- rownames(atac_da_s)

atac_cpm <- cpm(y_atac, log = TRUE)
atac_cpm_s <- atac_cpm[r_s, ]
atac_cpm_s <- atac_cpm_s[, c(grep(pattern = "PND15", x = colnames(atac_cpm_s)), grep(pattern = "Adult", x = colnames(atac_cpm_s)))]
```


# C
```{r}
tab_ge <- da[da$subtype_qvalue <= 0.05 & abs(da$subtype_logFC) >= 0.5, ]

load("input/repeat_elements_atac_diff_chromatin_accessibility.RData")

tab <- data.frame(rowData(data_te))
tab_e <- tab[tab$X.gene_id %in% tab_ge$gene_id, ]

rd <- tab_e
rownames(rd) <- rd$transcript_id
rd <- rd[abs(rd$logFC) >= 1 & rd$qvalue <= 0.05, ]
rd$Name <- "More accessible in adult spermatogonial cells"
rd$Name[rd$logFC < 0] <- "Less accessible in adult spermatogonial cells"
rd$Name <- factor(rd$Name, rev(unique(rd$Name)))

rd$anno <- as.character(rd$class)
# rd$anno[rd$anno == "intronic" | rd$anno == "exonic"] <- "gene_body"
rd$anno[rd$anno == "TSS" | rd$anno == "proximal <=1000bp"] <- "tss_prox_1kbp"

rd$anno[rd$anno == "exonic"] <- "Exonic"
rd$anno[rd$anno == "intronic"] <- "Intronic"
rd$anno[rd$anno == "intergenic"] <- "Intergenic"
rd$anno[rd$anno == "tss_prox_1kbp"] <- "TSS +/- 1kbp"

rd$anno <- factor(rd$anno, unique(rd$anno)[c(4,2,1,3)])

c <- ggplot(data = rd, mapping = aes(x = anno, fill = anno)) +
  geom_bar(position = "dodge", stat = "count") +
  geom_text(stat = "count", aes(label = ..count..), vjust = -0.1, size = 3, fontface = "bold") +
  scale_y_continuous(limits = c(0, 5000), breaks = c(0, 1000, 2000, 3000, 4000, 5000)) +
  fill_palette(palet = "Dark2") +
  # coord_flip() +
  facet_wrap(~Name, nrow = 2, labeller = label_wrap_gen(width = 25)) +
  ggtitle("") +
  xlab("") +
  ylab("Number of TE loci") +
  theme_bw(base_family = "Helvetica") +
  th +
  theme(
    legend.title = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    legend.key.height = unit(0, "mm"),
    legend.key.width = unit(3, "mm"),
    plot.margin = grid::unit(c(0, 0, 0, 0), "mm"),
    legend.text.align = 0
  ) +
  guides(fill = guide_legend(nrow = 2, byrow = F))

cg <- gridExtra::grid.arrange(egg::set_panel_size(p = c, width = unit(4, "cm"), height = unit(4, "cm")))

ggsave(plot = cg, filename = "output/c.pdf", width = unit(8.5, "in"), height = unit(11, "in"), dpi = 320)
```



# E
```{r}
# olf <- data.frame(read_xlsx("input/Olfr.genes_more.accessible.L1.targets.xlsx"))
load("input/data_pData.RData")

# ge <- olf$nearestTSS.gene_name

ge <- c(
  "Olfr1307", "Olfr646", "Olfr1116", "Olfr1116", "Olfr671", "Olfr362", "Olfr56", "Olfr56", "Olfr66",
  "Olfr1270", "Olfr1497", "Olfr418", "Olfr199", "Olfr53"
)


o_te <- tab[tab$nearestTSS.gene_name %in% ge, ]
o_te <- o_te[abs(o_te$distance2nearestTSS) <= 5000, ]

o_te <- o_te[abs(o_te$logFC) >= 1 & o_te$qvalue <= 0.05, ]
o_te <- o_te[o_te$X.family_id %in% "L1", ]

o_te <- o_te[, c("nearestTSS.gene_name", "logFC", "transcript_id", "gene_id")]

# o_te <- split(o_te, o_te$nearestTSS.gene_name)

# atac_olf <- ldply(lapply(o_te, function(x) mean(x[, 2])))
atac_olf <- o_te
colnames(atac_olf)[1:2] <- c("Genes", "logFC")
rownames(atac_olf) <- atac_olf$transcript_id

# y_atac <- DGEList(assay(data_te))
# y_atac <- calcNormFactors(y_atac)
#
# atac_cpm <- cpm(y_atac, log = TRUE)
# rownames(atac_cpm) <- data.frame(rowData(data_te))$transcript_id
# atac_cpm <- atac_cpm[o_te$X.transcript_id, ]
# atac_cpm <- data.frame(atac_cpm[,c(grep(pattern = "PND15", x = colnames(atac_cpm)), grep(pattern = "Adult", x = colnames(atac_cpm)))])
#
# mat <- atac_cpm
#
# mat <- data.matrix(mat - rowMeans(mat[, grep(pattern = "PND15", x = colnames(mat), value = T)]))
#
# o <- seriate(max(mat) - mat)
# o <- seriate(dist(mat), method = "MDS_angle")

# mat <- mat[get_order(o), ]
#
#
# Heatmap(mat, cluster_columns = F, cluster_rows = FALSE, split = o_te$nearestTSS.gene_name, row_title_rot = 0)+
#   rowAnnotation(GE = o_te$gene_id)


mat <- data$cpm[ge, ]
mat <- mat[, grep(pattern = "PND14|PNW8", x = colnames(mat))]
# mat <- data.matrix(mat - mat[, grep(pattern = "PND14", x = colnames(mat), value = T)])

o <- seriate(max(mat) - mat)
# o <- seriate(dist(mat), method = "MDS_angle")

mat <- mat[get_order(o), ]

cd <- data$pData[colnames(mat), 1, drop = FALSE]
cd$Group <- factor(cd$Group, levels = unique(cd$Group))

n_mat <- data.frame(Genes = rownames(mat), mat)
n_mat$Genes <- factor(n_mat$Genes, rownames(n_mat))

n_mat <- merge(n_mat, atac_olf)

n_mat <- n_mat[!duplicated(n_mat),]
rownames(n_mat) <- n_mat$transcript_id

# atac_olf <- atac_olf[rownames(mat), ]

# cls <- colorRamp2(c(0, 6), c("grey", "red"))

cls <- colorRamp2(breaks = c(-5, 0, 5), colors = c("blue", "white", "red"))




# set.seed(1887)
# e_a <- rowAnnotation(
#   df = atac_olf[, 2, drop = FALSE],
#   col = list(logFC = colorRamp2(c(0, 2), c("#F7FCB9", "#238443"))),
#   simple_anno_size = unit(3, "mm"),
#   show_annotation_name = TRUE,
#   # annotation_name_side = "right",
#   annotation_name_rot = 0,
#   annotation_name_gp = gpar(fontsize = 0, fontfamily = "Helvetica"),
#   annotation_legend_param = list(
#     title = expression(bold(~ Log[2] ~ "FC (ATAC-seq)")),
#     title_gp = gpar(fontsize = 8, fontface = "bold", fontfamily = "Helvetica"),
#     labels_gp = gpar(fontsize = 8, fontfamily = "Helvetica")
#   )
# )
# 
# e_a@anno_list$logFC@color_mapping@colors <- c(
#   e_a@anno_list$logFC@color_mapping@colors[1],
#   e_a@anno_list$logFC@color_mapping@colors[3],
#   e_a@anno_list$logFC@color_mapping@colors[5]
# )
# 
# e_a@anno_list$logFC@color_mapping@levels <- c(0, 1, 2)


ta <- HeatmapAnnotation(
    df = cd[, 1, drop = FALSE],
    col = list(Group = cols[3:4]),
    simple_anno_size = unit(3, "mm"),
    show_annotation_name = TRUE,
    annotation_name_side = "right",
    annotation_name_rot = 0,
    annotation_name_gp = gpar(fontsize = 0, fontfamily = "Helvetica"),
    annotation_legend_param = list(
      title_gp = gpar(fontsize = 8, fontface = "bold", fontfamily = "Helvetica"),
      labels_gp = gpar(fontsize = 8, fontfamily = "Helvetica")
    )
  )



# ra <- rowAnnotation(
#   df = n_mat[, 4, drop = FALSE],
#   col = list(logFC = colorRamp2(c(0, 2), c("#F7FCB9", "#238443"))),
#   simple_anno_size = unit(3, "mm"),
#   show_annotation_name = TRUE,
#   # annotation_name_side = "right",
#   annotation_name_rot = 0,
#   annotation_name_gp = gpar(fontsize = 0, fontfamily = "Helvetica"),
#   annotation_legend_param = list(
#     title = expression(bold(~ Log[2] ~ "FC (ATAC-seq)")),
#     title_gp = gpar(fontsize = 8, fontface = "bold", fontfamily = "Helvetica"),
#     labels_gp = gpar(fontsize = 8, fontfamily = "Helvetica")
#   )
# )
# 
# ra@anno_list$logFC@color_mapping@colors <- c(
#   ra@anno_list$logFC@color_mapping@colors[1],
#   ra@anno_list$logFC@color_mapping@colors[3],
#   ra@anno_list$logFC@color_mapping@colors[5]
# )
# 
# ra@anno_list$logFC@color_mapping@levels <- c(0, 1, 2)
# 



e1 <- Heatmap(
  matrix = n_mat[,2:3],
  # km = 2,
  col = hm_cols(x = n_mat[,2:3], col = rev(brewer.pal(n = 9, name = "RdBu")[c(1,5,9)])),
  cluster_rows = F,
  cluster_columns = F,
  clustering_method_columns = "ward.D2",
  show_row_names = F,
  show_column_names = F,
  border = T,
  name = "Enrichment (log)",
  row_split = n_mat$Genes,
  row_gap = unit(rep(0.5, 12), "mm"),
  column_split = cd[, 1, drop = FALSE],
  cluster_column_slices = T,
  show_parent_dend_line = FALSE,
  row_title_rot = 0, 
  row_title_gp = gpar(fontsize = 10, fontfamily = "Helvetica"),
  row_names_gp = gpar(fontsize = 10, fontfamily = "Helvetica"),
  column_title = NULL,
  column_dend_height = unit(0.5, "cm"),
  use_raster = FALSE,
  top_annotation = ta,
  # left_annotation = e_a,
  height = unit(x = 6, units = "cm"), width = unit(x = 2, units = "cm"),
  heatmap_legend_param = list(
    title = expression(bold(atop(Log[2]~CPM,(RNA-seq)))),
    legend_height = unit(1, "cm"),
    # title_position = "topcenter",
    title_gp = gpar(fontsize = 8, fontfamily = "Helvetica"),
    labels_gp = gpar(fontsize = 8, fontfamily = "Helvetica")
  )
) 


e2 <- Heatmap(
  matrix = n_mat[,4, drop = FALSE],
  # km = 2,
  col = colorRamp2(breaks = c(1, 1.25, 1.5, 1.75, 2), colors = brewer.pal(n = 5, name = "Blues")),
  cluster_rows = F,
  cluster_columns = F,
  clustering_method_columns = "ward.D2",
  show_row_names = T,
  show_column_names = F,
  # border = T,
  name = "ATAC",
  row_split = n_mat$Genes,
  row_gap = unit(rep(0.5, 12), "mm"),
  border  = T,
  # column_split = cd[, 1, drop = FALSE],
  # cluster_column_slices = T,
  show_parent_dend_line = FALSE,
  row_title_rot = 0, 
  row_title_gp = gpar(fontsize = 8, fontfamily = "Helvetica"),
  row_names_gp = gpar(fontsize = 8, fontfamily = "Helvetica"),
  column_title = NULL,
  column_dend_height = unit(0.5, "cm"),
  use_raster = FALSE,
  # top_annotation = ta,
  # left_annotation = e_a,
  height = unit(x = 6, units = "cm"), width = unit(x = 4, units = "mm"),
  heatmap_legend_param = list(
    title = expression(bold(atop(Log[2]~FC,(ATAC-seq)))),
    at  =c(1, 1.5, 2),
    legend_height = unit(1, "cm"),
    # title_position = "topcenter",
    title_gp = gpar(fontsize = 8, fontfamily = "Helvetica"),
    labels_gp = gpar(fontsize = 8, fontfamily = "Helvetica")
  )
)

e <- e1 + e2


# e


# set.seed(1100)
# e <- ta %v% Heatmap(
#   matrix = mat,
#   # km = 2,
#   col = cls,
#   cluster_rows = F,
#   cluster_columns = F,
#   clustering_method_columns = "ward.D2",
#   show_row_names = T,
#   show_column_names = F,
#   name = "Enrichment (log)",
#   # row_split = ge$Group,
#   column_split = cd[, 1, drop = FALSE],
#   cluster_column_slices = T,
#   show_parent_dend_line = FALSE,
#   row_title = NULL,
#   row_names_gp = gpar(fontsize = 8, fontfamily = "Helvetica"),
#   column_title = NULL,
#   column_dend_height = unit(0.5, "cm"),
#   use_raster = FALSE,
#   top_annotation = ta,
#   left_annotation = e_a,
#   height = unit(x = 6, units = "cm"), width = unit(x = 2, units = "cm"),
#   heatmap_legend_param = list(
#     title = expression(bold(~ Log[2] ~ "CPM (RNA-seq)")),
#     legend_height = unit(2, "cm"),
#     # title_position = "leftcenter-rot",
#     title_gp = gpar(fontsize = 8, fontfamily = "Helvetica"),
#     labels_gp = gpar(fontsize = 8, fontfamily = "Helvetica")
#   )
# )




pdf(file = "output/e.pdf", width = 8.5, height = 11)
draw(
  object = e,
  heatmap_legend_side = "right",
  annotation_legend_side = "right",
  merge_legends = TRUE,
  align_annotation_legend = "left"
)
dev.off()

png(filename = "output/e.png", width = 8.5, height = 11, units = "in", res = 330)
draw(
  object = e,
  heatmap_legend_side = "right",
  annotation_legend_side = "right",
  merge_legends = TRUE,
  align_annotation_legend = ""
)
dev.off()
```

# A 2
```{r}
ht_opt$TITLE_PADDING <- unit(1, "mm")
ht_opt$legend_gap <- unit(4, "mm")
ht_opt$legend_grid_height <- unit(3, "mm")
ht_opt$legend_grid_width <- unit(3, "mm")
ht_opt$HEATMAP_LEGEND_PADDING <- unit(1, "mm")

load("input/repeat_elements_atac_diff_chromatin_accessibility_gene_level.RData")

da <- data.frame(rowData(data_ge), stringsAsFactors = F, check.names = F)
atac_da <- da[da$subtype_qvalue <= 0.05 & da$subtype_logFC <= -0.5, ]
atac_da <- atac_da[atac_da$class_id %in% c("LINE", "LTR", "SINE"), ]
r <- rownames(atac_da)

y_atac <- DGEList(assay(data_ge))
colnames(y_atac) <- gsub(pattern = "input/|\\.bam", replacement = "", x = colnames(assay(data_ge)))
colnames(y_atac) <- gsub(pattern = "\\.", replacement = "_", x = colnames(y_atac))
y_atac <- calcNormFactors(y_atac)

atac_cpm <- cpm(y_atac, log = TRUE)
atac_cpm <- atac_cpm[r, ]
atac_cpm <- atac_cpm[, c(grep(pattern = "PND15", x = colnames(atac_cpm)), grep(pattern = "Adult", x = colnames(atac_cpm)))]

mat_atac <- data.matrix(atac_cpm - rowMeans(atac_cpm[, grep(pattern = "PND15", x = colnames(atac_cpm), value = T)]))
o_atac <- seriate(max(mat_atac) - mat_atac)
mat_atac <- mat_atac[get_order(o_atac), ]

cd <- data.frame(Samples = colnames(mat_atac))
cd$Group <- gsub(pattern = "_.*", replacement = "", x = cd$Samples)
rownames(cd) <- cd$Samples
cd <- cd[colnames(mat_atac), ]
cd$Group <- factor(cd$Group, levels = c("PND15", "Adult"))


ha <- HeatmapAnnotation(
  df = cd[, 2, drop = FALSE],
  col = list(Group = cols),
  simple_anno_size = unit(3, "mm"),
  show_annotation_name = TRUE,
  annotation_name_side = "right",
  annotation_name_rot = 0,
  annotation_name_gp = gpar(fontsize = 0, fontfamily = "Helvetica"),
  annotation_legend_param = list(
    title_gp = gpar(fontsize = 8, fontface = "bold", fontfamily = "Helvetica"),
    labels_gp = gpar(fontsize = 8, fontfamily = "Helvetica")
  )
)

colnames(atac_da)[1:3] <- c("Subtype", "Family", "Class")

rs <- c("L1", "ERVK", "ERVL", "ERV1", "ERVL-MaLR", "LINE", "LTR", "ERVK", "RTE-BovB", "ERV1", "ERVL-MaLR", "ERVL", "Gypsy")
rs <- unique(rs)
rc <- rcartocolor::carto_pal(n = 9, name = "Safe")
names(rc) <- rs

set.seed(1100)
a <- ha %v% Heatmap(
  matrix = mat_atac,
  col = hm_cols(x = mat_atac),
  cluster_rows = F,
  cluster_columns = T,
  clustering_method_columns = "ward.D2",
  show_row_names = F,
  show_column_names = F,
  row_gap = unit(0.5, "mm"),
  name = "Enrichment (log)",
  # row_split = ge$Group,
  column_split = cd[, 2, drop = FALSE],
  cluster_column_slices = T,
  show_parent_dend_line = FALSE,
  row_title = NULL,
  row_names_gp = gpar(fontsize = 8, fontfamily = "Helvetica"),
  column_title = NULL,
  column_dend_height = unit(0.5, "cm"),
  use_raster = FALSE,
  row_split = atac_da[,c(2,3)]  , row_title_rot = 0,
  top_annotation = ha,
  left_annotation = HeatmapAnnotation(
    df = atac_da[,c(3,2)],
    col = list(Class = rc, Family = rc),
    which = "row",
    simple_anno_size = unit(3, "mm"),
    show_annotation_name = TRUE,
    # annotation_name_side = "right",
    annotation_name_rot = 0,
    annotation_name_gp = gpar(fontsize = 0, fontfamily = "Helvetica"),
    annotation_legend_param = list(
      title_gp = gpar(fontsize = 8, fontface = "bold", fontfamily = "Helvetica"),
      labels_gp = gpar(fontsize = 8, fontfamily = "Helvetica")
    )
  ),
  height = unit(x = 6, units = "cm"), width = unit(x = 5, units = "cm"),
  heatmap_legend_param = list(
    title = expression(bold(~ Log[2] ~ "FC")),
    legend_height = unit(1, "cm"),
    at = c(0.5,0,-1,-2),
    # title_position = "leftcenter-rot",
    title_gp = gpar(fontsize = 8, fontfamily = "Helvetica"),
    labels_gp = gpar(fontsize = 8, fontfamily = "Helvetica")
  )
)


pdf(file = "output/a.pdf", width = 8.5, height = 11)
draw(
  object = a,
  heatmap_legend_side = "right",
  annotation_legend_side = "right",
  merge_legends = TRUE,
  align_annotation_legend = ""
)
dev.off()


png(filename = "output/a.png", width = 8.5, height = 11, units = "in", res = 330)
draw(
  object = a,
  heatmap_legend_side = "right",
  annotation_legend_side = "right",
  merge_legends = TRUE,
  align_annotation_legend = ""
)
dev.off()
```

# B 2
```{r}
load("input/limma_SC_Controls_lab_lit_tx.RData")

atac_da_s <- da[da$subtype_qvalue <= 0.05 & abs(da$subtype_logFC) >= 0.5, ]
ra <- merge(x = atac_da_s[, 1, drop = FALSE], y = dea.limma$`pnd14 vs pnw8`, all.x = T, by = "row.names")
rownames(ra) <- ra$Row.names

atac_da_s <- atac_da_s[atac_da_s$subtype_logFC > 0, ]
atac_da_s <- atac_da_s[atac_da_s$class_id %in% c("LINE", "LTR", "SINE"), ]
r_s <- rownames(atac_da_s)

atac_cpm <- cpm(y_atac, log = TRUE)
atac_cpm_s <- atac_cpm[r_s, ]
atac_cpm_s <- atac_cpm_s[, c(grep(pattern = "PND15", x = colnames(atac_cpm_s)), grep(pattern = "Adult", x = colnames(atac_cpm_s)))]

mat_atac_s <- data.matrix(atac_cpm_s - rowMeans(atac_cpm_s[, grep(pattern = "PND15", x = colnames(atac_cpm_s), value = T)]))
o_atac_s <- seriate(max(mat_atac_s) - mat_atac_s)
mat_atac_s <- mat_atac_s[get_order(o_atac_s), ]

ra <- ra[rownames(mat_atac_s), ]

colnames(atac_da_s)[1:3] <- c("Subtype", "Family", "Class")

set.seed(1001)
right_annotation <- HeatmapAnnotation(
  PND14vsPNW8 = ra$logFC,
  which = "row",
  simple_anno_size = unit(3, "mm"),
  show_annotation_name = TRUE,
  # annotation_name_side = "right",
  annotation_name_rot = 0,
  annotation_name_gp = gpar(fontsize = 0, fontfamily = "Helvetica"),
  annotation_legend_param = list(
    title = expression(bold(Log[2] ~ "FC (RNA-seq)")),
    title_gp = gpar(fontsize = 8, fontface = "bold", fontfamily = "Helvetica"),
    labels_gp = gpar(fontsize = 8, fontfamily = "Helvetica")
  )
)

right_annotation@anno_list$PND14vsPNW8@color_mapping@colors <- right_annotation@anno_list$PND14vsPNW8@color_mapping@colors[c(1, 6)]
right_annotation@anno_list$PND14vsPNW8@color_mapping@levels <- right_annotation@anno_list$PND14vsPNW8@color_mapping@levels[c(1, 6)]


ha1 <- HeatmapAnnotation(
  df = cd[, 2, drop = FALSE],
  col = list(Group = cols[c(2, 5)]),
  simple_anno_size = unit(3, "mm"),
  show_annotation_name = TRUE,
  annotation_name_side = "right",
  annotation_name_rot = 0,
  annotation_name_gp = gpar(fontsize = 0, fontfamily = "Helvetica"),
  annotation_legend_param = list(
    title_gp = gpar(fontsize = 8, fontface = "bold", fontfamily = "Helvetica"),
    labels_gp = gpar(fontsize = 8, fontfamily = "Helvetica")
  )
)


set.seed(1100)
b <- ha1 %v% Heatmap(
  matrix = mat_atac_s,
  col = hm_cols(mat_atac_s),
  cluster_rows = F,
  cluster_columns = T,
  clustering_method_columns = "ward.D2",
  show_row_names = F,
  show_column_names = F,
  row_gap = unit(0.5, "mm"),
  name = "Enrichment (log)",
  # row_split = ge$Group,
  column_split = cd[, 2, drop = FALSE],
  cluster_column_slices = T,
  show_parent_dend_line = FALSE,
  row_title = NULL,
  row_names_gp = gpar(fontsize = 8, fontfamily = "Helvetica"),
  column_title = NULL,
  column_dend_height = unit(0.5, "cm"),
  use_raster = FALSE,
  row_split = atac_da_s[,c(2,3)]  , row_title_rot = 0,
  left_annotation = HeatmapAnnotation(
    df = atac_da_s[,c(3,2)],
    col = list(Class = rc, Family = rc),
    which = "row",
    simple_anno_size = unit(3, "mm"),
    show_annotation_name = TRUE,
    # annotation_name_side = "right",
    annotation_name_rot = 0,
    annotation_name_gp = gpar(fontsize = 0, fontfamily = "Helvetica"),
    annotation_legend_param = list(
      title_gp = gpar(fontsize = 8, fontface = "bold", fontfamily = "Helvetica"),
      labels_gp = gpar(fontsize = 8, fontfamily = "Helvetica")
    )
  ), right_annotation = right_annotation,
  top_annotation = ha1,
  height = unit(x = 6, units = "cm"), width = unit(x = 5, units = "cm"),
  heatmap_legend_param = list(
    title = expression(bold(~ Log[2] ~ "FC (ATAC-seq)")),
    legend_height = unit(1, "cm"),
    at = c(-2, 0, 2),
    # title_position = "leftcenter-rot",
    title_gp = gpar(fontsize = 8, fontfamily = "Helvetica"),
    labels_gp = gpar(fontsize = 8, fontfamily = "Helvetica")
  )
)



pdf(file = "output/b.pdf", width = 8.5, height = 11)
draw(
  object = b,
  heatmap_legend_side = "right",
  annotation_legend_side = "right",
  merge_legends = TRUE,
  align_annotation_legend = ""
)
dev.off()

png(filename = "output/b.png", width = 8.5, height = 11, units = "in", res = 330)
draw(
  object = b,
  heatmap_legend_side = "right",
  annotation_legend_side = "right",
  merge_legends = TRUE,
  align_annotation_legend = ""
)
dev.off()
```


# References
```{r overlap-matrix-rGREAT-43 }
report::cite_packages(session = sessionInfo())
```

# SessionInfo
```{r overlap-matrix-rGREAT-44 }
devtools::session_info() %>%
  details::details()
```
