---
title: "ATAC-Seq/ RNA-Seq (PND15 vs Adults): Mapping TF Motifs to TFBA"
author: "Deepak Tanwar"
date: "<b>Created on:</b> 2019-11-29 <br> <b>Updated on:</b> `r Sys.Date()`"
output:
  rmdformats::readthedown:
    highlight: tango
    self_contained: true
    thumbnails: true
    lightbox: true
    gallery: false
    keep_md: no
    number_sections: no
    fig_width: 8
    fig_height: 8
    fig_caption: true
    df_print: paged
    code_folding: hide
  fontsize: 12pt
  geometry: margin=1in
  documentclass: article
# bibliography: references.bib
link-citations: yes
editor_options: 
  chunk_output_type: console
---

# Library
```{r 20191118-peaks-annotation-other-figs-1, message=FALSE, warning=FALSE}
library(plotly)
library(SummarizedExperiment)
```


# Data
```{r, warning=FALSE, message=FALSE}
atac <- readxl::read_xlsx(path = "input/atac_diff_acc_tfbs_hocomoco_overlap.xlsx", sheet = 1, trim_ws = T)

rna.all <- read.table("input/all_regulons_pnd15_adults.txt", header = T, sep = "\t", stringsAsFactors = F)
rna.all$"Transcription factor" <- rownames(rna.all)

rna.filt <- read.table("input/filt_regulons_pnd15_adults.txt", header = T, sep = "\t", stringsAsFactors = F)
rna.filt$"Transcription factor" <- rownames(rna.filt)
```


# Mapping
```{r}
atac.rna.all <- merge(atac, rna.all)
atac.rna.filt <- merge(atac, rna.filt)
```

# Save table
```{r}
write.table(
  x = atac.rna.all, file = "./output/tfMotifs_mapped_TFBA_all.txt",
  sep = "\t", quote = F, row.names = F
)
write.table(
  x = atac.rna.filt, file = "./output/tfMotifs_mapped_TFBA_filt.txt",
  sep = "\t", quote = F, row.names = F
)
```


# SessionInfo
```{r 20191118-peaks-annotation-other-figs-19 }
devtools::session_info()
```