---
title: "ATAC-Seq/ RNA-Seq (PND15 vs Adults): TF motifs overlap and diff exp genes"
author: "Deepak Tanwar"
date: "<b>Created on:</b> 2019-11-29 <br> <b>Updated on:</b> `r Sys.Date()`"
output:
  rmdformats::readthedown:
    highlight: tango
    self_contained: true
    thumbnails: true
    lightbox: true
    gallery: false
    keep_md: no
    number_sections: no
    fig_width: 8
    fig_height: 8
    fig_caption: true
    df_print: paged
    code_folding: hide
  fontsize: 12pt
  geometry: margin=1in
  documentclass: article
# bibliography: references.bib
link-citations: yes
editor_options: 
  chunk_output_type: console
---

# Library
```{r 20191129-map-tfMotifs-to-genes-1, message=FALSE, warning=FALSE}
library(plotly)
```


# Data
```{r 20191129-map-tfMotifs-to-genes-2, warning=FALSE, message=FALSE}
load("./input/limma_SC_Controls.RData")
dea <- dea.limma$`PND15 vs Adult`
colnames(dea) <- paste("DEA", colnames(dea), sep = "_")

dea <- dea[abs(dea$DEA_logFC) >= 1 & dea$DEA_adj.P.Val <= 0.05,]

tf <- readxl::read_xlsx(path = "input/atac_diff_acc_tfbs_hocomoco_overlap.xlsx", sheet = 1, trim_ws = T)
```


# Map with Genes
```{r 20191129-map-tfMotifs-to-genes-3}
tab <- merge(tf, dea, by.y = "DEA_Genes", by.x = "TF-Transcription factor")
```

# Save table
```{r 20191129-map-tfMotifs-to-genes-4}
writexl::write_xlsx(tab, "./output/dea_genes_overlap_tf_motifs.xlsx", col_names = T, format_headers = T)
```

# Plot for TF overlapping DEGs
```{r 20191129-map-tfMotifs-to-genes-5, fig.align='center', fig.height=7, fig.width=8, warning=FALSE, message=FALSE}
plot_ly(
  data = tab, x = tab$DEA_logFC, y = tab$`diffAccessibility-logFC`,
  marker = list(size = 5, color = "blue"),
  text = ~ paste(
    "Gene: ", tab$`Peak-SYMBOL`,
    "<br>Name: ", tab$`Peak-GENENAME`,
    "<br>Annotation: ", tab$`Peak-annotation`,
    "<br>DA logFC: ", tab$`diffAccessibility-logFC`,
    "<br>DA FDR: ", tab$`diffAccessibility-qvalue`,
    "<br>DE logFC: ", tab$DEA_logFC,
    "<br>DE FDR: ", tab$DEA_adj.P.Val
  ),
  hoverinfo = "text"
) %>%
  layout(
    title = "TF Motifs overlapping differentially expressed genes",
    yaxis = list(title = "logFC: Differential Accessibility (DA)"),
    xaxis = list(title = "logFC: Differential Expression (DE)")
  )
```


# SessionInfo
```{r 20191129-map-tfMotifs-to-genes-6 }
devtools::session_info()
```