---
title: "DO: Enriched Heatmaps"
author: "Deepak Tanwar"
date: "<b>Created on:</b> 2020-07-31 <br> <b>Updated on:</b> `r Sys.Date()`"
output:
  rmdformats::readthedown:
    highlight: tango
    self_contained: true
    thumbnails: true
    lightbox: true
    gallery: false
    keep_md: no
    number_sections: no
    fig_width: 8
    fig_height: 8
    fig_caption: true
    df_print: paged
    code_folding: show
  fontsize: 12pt
  geometry: margin=1in
  documentclass: article
# bibliography: references.bib
link-citations: yes
---

# Library
```{r integration-all-1, warning = F, message = F}
library(EnrichedHeatmap)
library(circlize)
library(data.table)
library(tidyverse)
library(plgINS)
library(BSgenome.Mmusculus.UCSC.mm10)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(csaw)
library(gUtils)
library(report)
library(plyr)
```


# Function to subset data to one chromosome
```{r integration-all-3, echo = F}
mylapply <- function(...) {
  if (require(parallel) && .Platform$OS.type == "unix") {
    mclapply(..., mc.preschedule = F)
  }
  else {
    lapply(...)
  }
}
```

# Data overlaps
```{r}
load("input/overlap_tables.RData")
wo <- ldply(sp_df_s1, data.frame)[, -1]
# wi <- ldply(sp_df_s2, data.frame)[, -1]
```

# ATAC differential analysis
```{r integration-all-4 }
load("input/atac_diff_chromatin_accessibility.RData")

res <- data.frame(rowData(data), stringsAsFactors = F, check.names = F)
colnames(res) <- gsub(pattern = "Peak-", replacement = "", x = colnames(res))
res_p <- res[res$`diffAccessibility-qvalue` <= 0.05, ]
res_lp <- res[res$`diffAccessibility-qvalue` <= 0.05 & abs(res$`diffAccessibility-logFC`) > 1, ]
res_lp <- inner_join(res_lp, wo)


tss_mm10 <- gr.mid(GRanges(res_lp))
# tss_mm10 <- one_chr(tss_mm10)
names(tss_mm10) <- tss_mm10$Name

tss_5k <- promoters(tss_mm10, upstream = 2500, downstream = 2500)
# tss_5k <- one_chr(tss_5k)
names(tss_5k) <- tss_5k$Name
```

# Normalize signal to tss

## ATAC-Seq
```{r integration-all-5 }
nm_atac <- "./output/mat_atac.RData"

if (!file.exists(nm_atac)) {
  atac_files <- list.files("input/atacseq", pattern = "\\.bw$", full.names = TRUE)
  names(atac_files) <- gsub(pattern = "\\.bw", replacement = "", x = basename(atac_files))
  atac_bw <- mylapply(atac_files, function(x) rtracklayer::import(x))

  mat_AS <- mylapply(atac_bw, FUN = function(x) {
    normalizeToMatrix(x, tss_mm10,
      extend = 2500,
      value_column = "score",
      include_target = TRUE,
      mean_mode = "w0",
      w = 50
    )
  }, mc.cores = length(atac_files))

  save(mat_AS, file = nm_atac)
} else {
  load(nm_atac)
}
```


## ChIP-Seq
```{r integration-all-6 }
nm_chip <- "./output/mat_chip.RData"

if (!file.exists(nm_chip)) {
  chip_files <- list.files("input/chipseq", pattern = "\\.bw$", full.names = TRUE)
  names(chip_files) <- gsub(pattern = "\\.bw", replacement = "", x = basename(chip_files))
  chip_bw <- mylapply(chip_files, function(x) rtracklayer::import(x))

  mat_CS <- mylapply(chip_bw, FUN = function(x) {
    normalizeToMatrix(x, tss_mm10,
      extend = 2500,
      value_column = "score",
      include_target = TRUE,
      mean_mode = "w0",
      w = 50
    )
  }, mc.cores = length(chip_files))

  save(mat_CS, file = nm_chip)
} else {
  load(nm_chip)
}
```


## Bisulphite-Seq
```{r integration-all-7 }
nm_bs <- "./output/mat_bs.RData"

if (!file.exists(nm_bs)) {
  bs_files <- list.files("input/bsseq", pattern = "\\.gz$", full.names = TRUE)
  names(bs_files) <- gsub(
    pattern = "_1_val_1_bismark_bt2_pe.bismark.cov.gz",
    replacement = "", x = basename(bs_files)
  )

  bs_cov <- mylapply(bs_files, function(x) {
    fread(
      input = x,
      sep = "\t", quote = F, stringsAsFactors = F,
      data.table = FALSE, nThread = detectCores(), showProgress = F,
      col.names = c("seqnames", "start", "end", "percent", "Me", "Un")
    )
  }, mc.cores = length(bs_files))

  bs_cov <- mylapply(bs_cov, function(x) {
    # a <- one_chr(GRanges(x))
    a <- GRanges(x)
    a$meth <- a$percent / 100
    a <- a[, 4]
    return(a)
  }, mc.cores = length(bs_files))


  mat_BS <- mylapply(bs_cov, FUN = function(x) {
    normalizeToMatrix(x, tss_mm10,
      extend = 2500,
      value_column = "meth",
      include_target = TRUE,
      smooth = TRUE,
      mean_mode = "absolute",
      background = NA
    )
  }, mc.cores = length(bs_files))

  save(mat_BS, file = nm_bs)
} else {
  load(nm_bs)
}
```


## RRBS
```{r integration-all-8 }
nm_rrbs <- "./output/mat_rrbs.RData"

if (!file.exists(nm_rrbs)) {
  rrbs_file <- "input/rrbs/PND8_trimmed_bismark_bt2.bismark.cov.gz"
  names(rrbs_file) <- gsub(
    pattern = "_trimmed_bismark_bt2.bismark.cov.gz",
    replacement = "", x = basename(rrbs_file)
  )

  rrbs_cov <- fread(
    input = rrbs_file,
    sep = "\t", quote = F, stringsAsFactors = F,
    data.table = FALSE, nThread = detectCores(), showProgress = F,
    col.names = c("seqnames", "start", "end", "percent", "Me", "Un")
  )

  # rrbs_cov <- one_chr(GRanges(rrbs_cov))
  rrbs_cov <- GRanges(rrbs_cov)
  rrbs_cov$meth <- rrbs_cov$percent / 100
  rrbs_cov <- rrbs_cov[, 4]

  mat_RRBS <- normalizeToMatrix(
    signal = rrbs_cov, target = tss_mm10,
    extend = 2500,
    value_column = "meth",
    include_target = TRUE,
    smooth = TRUE,
    mean_mode = "absolute",
    background = NA
  )

  save(mat_RRBS, file = nm_rrbs)
} else {
  load(nm_rrbs)
}
```

## RNA-Seq

```{r integration-all-9 }
nm_rna_counts <- "./output/mat_rna_counts.RData"

if (!file.exists(nm_rna_counts)) {
  load("input/data_pData.RData")
  cpm <- data$cpm

  tmp1 <- data.frame(Name = tss_5k$Name, Gene = tss_5k$nearestTSSgene_name)
  tmp2 <- data.frame(Gene = rownames(cpm), cpm)
  tab <- left_join(tmp1, tmp2)
  rownames(tab) <- tab$Name
  mat_RNA_counts <- tab[rownames(mat_AS$Adult_NF), -c(1:2)]

  mat_RNA_counts <- mat_RNA_counts - rowMeans(mat_RNA_counts[, grep(
    pattern = "PND8", x = colnames(mat_RNA_counts),
    value = T
  )])

  mat_RNA_counts_merged <- data.frame(
    Adult = rowMeans(mat_RNA_counts[, grep(pattern = "Adult", x = colnames(mat_RNA_counts))]),
    PND15 = rowMeans(mat_RNA_counts[, grep(pattern = "PND15", x = colnames(mat_RNA_counts))]),
    PND8 = rowMeans(mat_RNA_counts[, grep(pattern = "PND8", x = colnames(mat_RNA_counts))])
  )

  save(mat_RNA_counts, mat_RNA_counts_merged, file = nm_rna_counts)
} else {
  load(nm_rna_counts)
}
```



# Heatmaps

## Function
```{r integration-all-10 }
generate_diff_color_fun <- function(x) {
  q <- quantile(x, c(0.5, .95))
  max_q <- max(abs(q))
  colorRamp2(c(-max_q, 0, max_q), c("#3794bf", "#FFFFFF", "#df8640"))
}

generate_color_fun <- function(x, col = "red") {
  q <- NULL
  if (is.list(x)) {
    a <- lapply(x, function(y) quantile(y, c(0.05, 0.95)))
    min <- NULL
    max <- NULL
    for (i in 1:length(a)) {
      min <- c(min, a[[i]][1])
      max <- c(max, a[[i]][2])
    }
    q <- c(min(min), max(max))
  } else {
    q <- quantile(x, c(0.05, 0.95))
  }
  max_q <- max(abs(q))
  col <- colorRamp2(c(0, max_q), c("#FFFFFF", col))
  return(col)
}

create_heatmap <- function(split = T) {
  set.seed(1100)

  sp <- NULL

  if (split) {
    sa <- res_lp[, c("anno", "Name")]
    sa <- split(x = sa, f = sa$anno)
    sa1 <- sa[lapply(sa, nrow) > 500]
    sa2 <- sa[lapply(sa, nrow) < 500]
    sa2 <- plyr::ldply(sa2, data.frame)[,-1]
    sp <- sa1
    sp[["others"]] <- sa2
  } else {
    sp <- list(res_lp[, c("anno", "Name")])
    names(sp) <- "all"
  }
  

  for (i in 1:length(sp)) {
    x <- sp[[i]]

    k <- x$anno
    names(k) <- x$Name

    partition <- k

    nc <- length(unique(k))

    pdf_file <- paste0("./output/heatmap_atac_chip_", names(sp)[i], ".pdf")

    if (!file.exists(pdf_file)) {
      col_an <- RColorBrewer::brewer.pal(n = 8, name = "Dark2")
      names(col_an) <- unique(partition)

      col_an <- col_an[!is.na(names(col_an))]

      hm_an <- HeatmapAnnotation(
        Groups = partition, which = "row",
        Groups = col_an
      )

      axis_name <- c("-2.5kb", "Peak-mid", "2.5kb")

      # an_atac <- HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:(nc + 1)), ylim = c(0, 5)))

      col_atac <- generate_color_fun(mat_AS)

      # ATAC-Seq
      hm_atac <- EnrichedHeatmap(mat_AS$Adult_NF[x$Name, ],
        name = "Adult ATAC",
        col = col_atac,
        column_title = "Adult ATAC",
        # top_annotation = an_atac
        axis_name = axis_name,
        top_annotation = HeatmapAnnotation(lines = anno_enriched(
          gp = gpar(col = 2:(nc + 1))
          # ylim = c(0, 55)
        )),
        use_raster = FALSE,
        row_split = partition,
        row_title = NULL
      ) +
        EnrichedHeatmap(mat_AS$PND15_NF[x$Name, ],
          name = "PND15 ATAC",
          col = col_atac,
          column_title = "PND15 ATAC",
          # top_annotation = an_atac
          axis_name = axis_name,
          top_annotation = HeatmapAnnotation(lines = anno_enriched(
            gp = gpar(col = 2:(nc + 1)),
            # ylim = c(0, 55)
          )),
          use_raster = FALSE,
          row_split = partition,
          row_title = NULL
        )


      # ChIP-Seq
      hm_chip <- EnrichedHeatmap(mat_CS$PNW8_H3K4me3_SRX332343[x$Name, ],
        name = "Adult H3K4me3",
        col = generate_color_fun(mat_CS$PNW8_H3K4me3_SRX332343[x$Name, ], col = "blue"),
        column_title = "Adult H3K4me3",
        axis_name = axis_name,
        top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:(nc + 1)))),
        use_raster = FALSE,
        row_split = partition,
          row_title = NULL
      ) +
        EnrichedHeatmap(mat_CS$PNW8_H3K27me3_SRX332346[x$Name, ],
          name = "Adult H3K27me3",
          col = generate_color_fun(mat_CS$PNW8_H3K27me3_SRX332346[x$Name, ], col = "blue"),
          column_title = "Adult H3K27me3",
          axis_name = axis_name,
          top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:(nc + 1)))),
          use_raster = FALSE,
          row_split = partition,
          row_title = NULL
        ) +
        EnrichedHeatmap(mat_CS$PNW8_H3K27ac_SRX332351[x$Name, ],
          name = "Adult H3K27ac",
          col = generate_color_fun(mat_CS$PNW8_H3K27ac_SRX332351[x$Name, ], col = "blue"),
          column_title = "Adult H3K27ac",
          axis_name = axis_name,
          top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:(nc + 1)))),
          use_raster = FALSE,
          row_split = partition,
          row_title = NULL
        )


      # Methylation color
      meth_col_fun <- colorRamp2(c(0, 0.5, 1), c("red", "white", "blue"))

      # BSseq
      hm_bs <- EnrichedHeatmap(mat_BS$PND7_SRX749887[x$Name, ],
        name = "PND7 BS",
        # col = c("white", "#ba1168"),
        col = meth_col_fun,
        column_title = "PND7 BS",
        axis_name = axis_name,
        top_annotation = HeatmapAnnotation(lines = anno_enriched(
          gp = gpar(col = 2:(nc + 1)),
          # ylim = c(0, 3)
        )),
        use_raster = FALSE,
        row_split = partition,
          row_title = NULL
      ) +
        EnrichedHeatmap(mat_BS$PND14_SRX749892[x$Name, ],
          name = "PND14 BS",
          # col = c("white", "#ba1168"),
          col = meth_col_fun,
          column_title = "PND14 BS",
          axis_name = axis_name,
          top_annotation = HeatmapAnnotation(lines = anno_enriched(
            gp = gpar(col = 2:(nc + 1)),
            # ylim = c(0, 3)
          )),
          use_raster = FALSE,
          row_split = partition,
          row_title = NULL
        ) +
        EnrichedHeatmap(mat_BS$PNW8[x$Name, ],
          name = "PNW8 BS",
          # col = c("white", "#ba1168"),
          col = meth_col_fun,
          column_title = "PNW8 BS",
          axis_name = axis_name,
          top_annotation = HeatmapAnnotation(lines = anno_enriched(
            gp = gpar(col = 2:(nc + 1)),
            # ylim = c(0, 3)
          )),
          use_raster = FALSE,
          row_split = partition,
          row_title = NULL
        )


      # # RRBS
      # hm_rrbs <- EnrichedHeatmap(mat_RRBS[x$Name, ],
      #   name = "PND8 RRBS",
      #   # col = c("white", "#ba1168"),
      #   column_title = "PND8 RRBS",
      #   col = meth_col_fun,
      #   axis_name = axis_name,
      #   top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:(nc + 1)))),
      #   use_raster = FALSE,
      # row_split = partition,
      #     row_title = NULL
      # )


      ## Counts

      an_RNA <- data.frame(Group = gsub(pattern = "_.*", replacement = "", x = colnames(mat_RNA_counts)))
      rownames(an_RNA) <- colnames(mat_RNA_counts)

      an_RNA_m <- data.frame(Group = colnames(mat_RNA_counts_merged))
      rownames(an_RNA_m) <- an_RNA_m$Group

      try(
      hm_rna_counts <- Heatmap(
        matrix = data.matrix(mat_RNA_counts[x$Name, ]),
        name = "RNA Expression", na_col = "grey",
        show_row_names = FALSE,
        show_column_names = FALSE,
        cluster_rows = F, cluster_columns = F,
        top_annotation = HeatmapAnnotation(df = an_RNA),
        width = unit(5, "cm"), use_raster = FALSE,
        row_split = partition,
          row_title = NULL
      # ) +
      #   Heatmap(
      #     matrix = data.matrix(mat_RNA_counts_merged[x$Name, ]),
      #     name = "RNA Expression (mean)", na_col = "grey",
      #     show_row_names = FALSE,
      #     show_column_names = FALSE,
      #     cluster_rows = F, cluster_columns = F,
      #     top_annotation = HeatmapAnnotation(df = an_RNA_m),
      #     width = unit(2, "cm"), use_raster = FALSE,
      # row_split = partition,
      #     row_title = NULL
        )
)

      pdf(file = pdf_file, width = 16.5, height = 12.75)

      if (nc == 1) {
        ht_list <- hm_an + hm_atac + hm_chip + hm_bs + hm_rna_counts
        draw(ht_list, ht_gap = unit(2, "mm"))
      } else {
        ht_list <- hm_an + hm_atac + hm_chip + hm_bs + hm_rna_counts
        draw(ht_list,
          # row_split = partition,
          # annotation_legend_list = list(lgd),
          ht_gap = unit(2, "mm"),
          # row_title = NULL
        )
      }

      dev.off()
      invisible(gc())
    }
  }
}
```

## Make heatmaps
```{r integration-all-11 }
create_heatmap(split = FALSE)
create_heatmap(split = TRUE)
```

## Convert to PNG
```{r integration-all-12, eval = TRUE}
pdf <- list.files(path = "output", pattern = "pdf", recursive = T, full.names = T)
png <- mylapply(pdf, function(x) {
  pdftools::pdf_convert(
    pdf = x, format = "png", dpi = 200,
    filenames = gsub(pattern = ".pdf", replacement = ".png", x = x), pages = 1
  )
}, mc.cores = length(pdf))
```

# Heatmaps
```{r integration-all-13, echo=FALSE}
plots <- list.files(path = "output", pattern = "png", full.names = TRUE)
plots <- plots[grep(pattern = "heat", x = plots)]

names(plots) <- sapply(plots, function(x) {
  n <- as.numeric(gsub(pattern = "\\..*|.*_cluster_", replacement = "", x = x))
  if (n == 1) {
    return("No clustering")
  } else {
    return(paste("Cluster:", sprintf("%01d", n)))
  }
})

plots <- plots[c(2:10, 1)]

bs <- bsselectR::bsselect(plots, type = "img", live_search = TRUE, show_tick = TRUE, height = 1000, width = "100%")

bsselectR::as_iframe(widget = bs, file = "lfc1_qval05.html", height = 1000, width = "100%")
```

# SessionInfo
```{r integration-all-14 }
devtools::session_info()
```
