---
title: "Repetitive elements in ATAC-Seq: Differential analysis"
author: "Deepak Tanwar"
date: "<b>Created on:</b> 2020-01-29 <br> <b>Updated on:</b> `r Sys.Date()`"
output:
  rmdformats::readthedown:
    highlight: tango
    self_contained: true
    thumbnails: true
    lightbox: true
    gallery: false
    keep_md: no
    number_sections: no
    fig_width: 8
    fig_height: 8
    fig_caption: true
    df_print: paged
    code_folding: show
  fontsize: 12pt
  geometry: margin=1in
  documentclass: article
# bibliography: references.bib
link-citations: yes
---

# Library
```{r, message=FALSE, warning=FALSE}
library(csaw)
library(edgeR)
library(Rsubread)
```

# Reading the counts

```{r, message = F, warning = F}
# BAM files
bamFiles <- list.files(path = "input", pattern = "bam$", full.names = T)
names(bamFiles) <- gsub(pattern = "\\.bam", replacement = "", basename(bamFiles))
gtf <- data.frame(rtracklayer::import("input/mm10_rmsk_TE.gtf.gz"))

rtracklayer::export(object = gtf[1:100,],con = "test.gtf", format = "gtf")

# gtf$Derives_from <- paste(gtf$class_id, gtf$family_id, sep = "/")
gtf1 <- data.frame(rtracklayer::import("/mnt/IM/anno/Mus_musculus.GRCm38.75.repeats.gtf"))
# gtf_test <- gtf[sample(nrow(gtf), 1000), ]
```

## Family
```{r, warning=FALSE, message=FALSE}
fc_family <- featureCounts(
  files = bamFiles,
  annot.ext = "input/mm10_rmsk_TE.gtf.gz",
  isGTFAnnotationFile = T,
  GTF.attrType = "class_id",
  GTF.featureType = "exon",
  allowMultiOverlap = T,
  fraction = T,
  countMultiMappingReads = T,
  ignoreDup = T,
  nthreads = 16
)

saveRDS(object = fc_family, file = "./output/fc_Family.rds")

rownames(fc_family$annotation) <- fc_family$annotation$GeneID
counts_family <- SummarizedExperiment(list(counts = fc_family$counts))
counts_family@metadata <- fc_family$annotation
```


## subFamily
```{r, warning=FALSE, message=FALSE}
fc_sub_family <- featureCounts(
  isPairedEnd = TRUE,
  files = bamFiles,
  annot.ext = "input/mm10_rmsk_TE.gtf.gz",
  isGTFAnnotationFile = T,
  GTF.attrType = "class_id",
  GTF.attrType.extra = "family_id",
  GTF.featureType = "exon",
  allowMultiOverlap = T,
  fraction = T,
  countMultiMappingReads = T,
  ignoreDup = T,
  nthreads = 16
)


saveRDS(object = fc_sub_family, file = "./output/fc_subFamily.rds")

rownames(fc_sub_family$annotation) <- fc_sub_family$annotation$GeneID
counts_sub_family <- SummarizedExperiment(list(counts = fc_sub_family$counts))
counts_sub_family@metadata <- fc_sub_family$annotation
```



# Filtering counts

## Function
```{r}
counts <- readRDS("output/counts.rds")

counts_filter <- function(counts, nReads = 15, nSamplesPercent = 0.4) {
  counts2 <- apply(counts, 1, FUN = function(x) {
    sum(x >= 15) >= floor(length(x) * 0.4)
  })
  return(counts2)
}
```

## Filter counts
```{r}
keep <- counts_filter(counts = assay(counts))
counts.filt <- counts[keep]
counts.filt$totals <- colSums(assay(counts.filt))
```


# Differential binding
```{r}
data <- normFactors(counts.filt)
# colnames(rowData(data)) <- paste("Peak", colnames(rowData(data)), sep = "-")

y <- asDGEList(data)
colnames(y) <- names(bamFiles)

design <- model.matrix(~ 0 + factor(gsub("_.*", "", colnames(y))))
colnames(design) <- gsub(".*)", "", colnames(design))

contrast.matrix <- makeContrasts(Adult - PND15, levels = design)

y <- estimateDisp(y, design)
fit <- glmQLFit(y, design, robust = TRUE)
results <- glmQLFTest(fit, contrast = contrast.matrix)
results$table$qvalue <- p.adjust(p = results$table$PValue, method = "BH")
# colnames(results$table) <- paste("diffEnrichment", colnames(results$table), sep = "-")
rowData(data) <- cbind(rowData(data), results$table)

tab <- data.frame(rowData(data))
tab <- data.frame(Name = rownames(tab), tab)
tab <- merge(gtf[,10:11], tab)

rowData(data) <- tab
```


# Saving results
```{r}
save(data, file = "output/repeat_elements_atac_diff_chromatin_accessibility.RData")

tab <- data.frame(rowData(data), check.names = F, stringsAsFactors = F)

data.table::fwrite(
  x = tab,
  file = "./output/repeat_elements_atac_diff_chromatin_accessibility.txt.gz",
  sep = "\t", quote = F, row.names = F, nThread = 16
)

tab.sig <- tab[abs(tab$logFC) >= 0.5 & tab$qvalue <= 0.05, ]

writexl::write_xlsx(
  x = tab.sig, path = "./output/repeat_elements_atac_diff_chromatin_accessibility_sig.xlsx",
  col_names = T, format_headers = T
)
```


```{r}
t <- readxl::read_xlsx("output/repeat_elements_atac_diff_chromatin_accessibility_sig.xlsx")
```


<!-- # Merging with diff Accessibility -->
<!-- ```{r} -->
<!-- tab1 <- rowData(data) -->
<!-- load("input/atac_diff_chromatin_accessibility.RData") -->
<!-- tab2 <- rowData(data) -->
<!-- tab2 <- tab2[tab2$`diffAccessibility-qvalue` <= 0.05 & abs(tab2$`diffAccessibility-logFC`) >= 1, ] -->
<!-- colnames(tab2)[1:5] <- gsub(pattern = "Peak-", replacement = "", x = colnames(tab2)[1:5]) -->

<!-- tab1_GR <- GRanges(tab1) -->
<!-- tab2_GR <- GRanges(tab2) -->

<!-- meta <- data.frame(plgINS::metaGR(gr1 = tab1_GR, gr2 = tab2_GR, minOverlap = 5), -->
<!--   stringsAsFactors = F -->
<!-- ) -->
<!-- colnames(meta)[1:13] <- paste0("RE-", colnames(meta)[1:13]) -->
<!-- colnames(meta) <- gsub(pattern = "\\.", replacement = "-", x = colnames(meta)) -->
<!-- colnames(meta) <- gsub(pattern = "GRanges-", replacement = "", x = colnames(meta)) -->
<!-- colnames(meta)[19:23] <- paste0("Peak-", colnames(meta)[19:23]) -->

<!-- writexl::write_xlsx( -->
<!--   x = meta, path = "output/repeat_elements_overlap_peaks.xlsx", -->
<!--   col_names = T, format_headers = T -->
<!-- ) -->
<!-- ``` -->


# Annotation
```{r}
load("output/repeat_elements_atac_diff_chromatin_accessibility.RData")

rd <- GRanges(rowData(data))

pa <- plgINS:::annoPeaks(
  peaks = rd,
  gtf = "/mnt/IM/reference/annotation/gencodeM18/gencode.vM18.chr_patch_hapl_scaff.annotation.gtf.gz", 
  proximal = 1000)

pa$class_n <- as.character(pa$class)
pa$class_n[pa$class_n == "intronic"] <- "gene_body"
pa$class_n[pa$class_n == "exonic"] <- "gene_body"
pa$class_n[pa$class_n == "TSS"] <- "Tss_prox_1kbp"
pa$class_n[pa$class_n == "proximal <=1000bp"] <- "Tss_prox_1kbp"

pa <- data.frame(pa, check.names = F, stringsAsFactors = F)
pa <- pa[, c(grep(pattern = "diffEnri", x = colnames(pa), invert = T), grep(pattern = "diffEnri", x = colnames(pa)))]

colnames(pa) <- gsub(pattern = "\\.", replacement = "-", x = colnames(pa))

rowRanges(data) <- GRanges(pa)
rowData(data) <- DataFrame(pa)

save(data, file = "output/repeat_elements_atac_diff_chromatin_accessibility.RData")

tab <- data.frame(rowData(data), check.names = F, stringsAsFactors = F)

data.table::fwrite(
  x = tab,
  file = "./output/repeat_elements_atac_diff_chromatin_accessibility.txt.gz",
  sep = "\t", quote = F, row.names = F, nThread = 16
)

tab.sig <- tab[abs(tab$diffEnrichment.logFC) >= 1 & tab$diffEnrichment.qvalue <= 0.05, ]

writexl::write_xlsx(
  x = tab.sig, path = "./output/repeat_elements_atac_diff_chromatin_accessibility_sig.xlsx",
  col_names = T, format_headers = T
)
```

# Adding RNA expression data
```{r}
re <- readxl::read_xlsx("output/repeat_elements_atac_diff_chromatin_accessibility_sig.xlsx")

load("input/limma_SC_Controls.RData")

de <- dea.limma$`PND15 vs Adult`

re_de <- data.frame(dplyr::left_join(x = re, y = de, by = c("nearestTSS.gene_name" = "Genes")))

writexl::write_xlsx(
  x = re_de, path = "./output/repeat_elements_atac_diff_chromatin_accessibility_sig_merge_diff_expression.xlsx",
  col_names = T, format_headers = T
)

o <- re_de[!is.na(re_de$logFC),]

plot(o$diffEnrichment.logFC, o$logFC)
```


#################################################################################################################

# Perform DA at Family and SubFamily level

```{r}
tab <- assay(counts)
colnames(tab) <- gsub(pattern = "input/|\\.bam", replacement = "", x = counts@colData$bam.files)

rd <- data.frame(rowData(counts)[,10:13])
rd$sf <- paste(rd$class_id, rd$family_id, sep = "/")

rownames(rd) <- rd$transcript_id

# t <- readxl::read_xlsx("output/repeat_elements_atac_diff_chromatin_accessibility_sig.xlsx")
# rd <- rd[t$transcript_id,]


sf <- data.frame(cbind(ID = rd[,"sf"], tab), stringsAsFactors = F)



sf <- plgINS::plag(x = data.matrix(sf[,2:ncol(sf)]), by = sf$ID)

sf_se <- SummarizedExperiment(assays = list(counts = data.matrix(sf)), colData = counts@colData[,1, drop = FALSE])
```

```{r}
# perform_da <- function(se){
# # keep <- counts_filter(counts = assay(se))
# # se_filt <- se[keep]

se$totals <- colSums(assay(se))
data <- normFactors(se)

y <- asDGEList(data)
# colnames(y) <- names(bamFiles)

# design <- model.matrix(~ 0 + factor(gsub("_.*", "", colnames(y))))
# colnames(design) <- gsub(".*)", "", colnames(design))
# 
# contrast.matrix <- makeContrasts(Adult - PND15, levels = design)
# 
# y <- estimateDisp(y, design)
# fit <- glmQLFit(y, design, robust = TRUE)
# results <- glmQLFTest(fit, contrast = contrast.matrix)
# results$table$qvalue <- p.adjust(p = results$table$PValue, method = "BH")
# # colnames(results$table) <- paste("diffEnrichment", colnames(results$table), sep = "-")
# rowData(data) <- cbind(rowData(data), results$table)
# 
# tab <- data.frame(rowData(data))
# tab <- data.frame(Name = rownames(tab), tab)
# # tab <- merge(gtf[,10:11], tab)
# 
# rowData(data) <- tab
# }
```




# Heatmap
```{r}
library(viridis)
cols <- magma(n = 10)[c(3, 6, 9)]
cols <- c(PND8 = cols[3], PND15 = cols[2], Adult = cols[1])

re_ele <- cpm(y, normalized.lib.sizes = T, log = T)

re_ele <- data.frame(symbol = rownames(re_ele), re_ele)
re_ele <- re_ele[grep(pattern = "\\?", x = re_ele$symbol, value = T, invert = T), ]
re_ele <- re_ele[grep(pattern = "LTR|LINE|SINE", x = rownames(re_ele)), ]
mat <- re_ele[,-1]
mat <- data.matrix(mat - rowMeans(mat[, grep(pattern = "PND15", x = colnames(mat), value = T)]))

library(ComplexHeatmap)
library(seriation)

o <- seriate(max(mat) - mat)

mat <- mat[get_order(o), ]

rd <- data.frame(ID = rownames(mat))
rd$Category <- gsub(pattern = "\\/.*", replacement = "", x = rd$ID)
rownames(rd) <- rd$ID

rd <- rd[rownames(mat),]

cd <- data.frame(ID = colnames(mat))
cd$Group <- gsub(pattern = "_.*", replacement = "", x = cd$ID)

cd$Group <- factor(cd$Group, levels = rev(unique(cd$Group)))
rownames(cd) <- cd$ID

ele_cols <- RColorBrewer::brewer.pal(n = 3, name = "Dark2")
names(ele_cols) <- c("LINE", "LTR", "SINE")

la <- HeatmapAnnotation(
  Category = rd$Category, which = "row",
  col = list(Category = ele_cols),
  simple_anno_size = unit(3, "mm"),
  show_annotation_name = TRUE,
  annotation_name_side = "bottom",
  annotation_legend_param = list(
    title_gp = gpar(fontsize = 8, fontface = "bold", fontfamily = "Helvetica"),
    labels_gp = gpar(fontsize = 8, fontfamily = "Helvetica")
  ),
  annotation_name_rot = 0,
  annotation_name_gp = gpar(fontsize = 0, fontfamily = "Helvetica")
)

ha <- HeatmapAnnotation(
  df = cd[, 2, drop = FALSE],
  col = list(Group = cols[2:3]),
  simple_anno_size = unit(3, "mm"),
  show_annotation_name = TRUE,
  annotation_name_side = "right",
  annotation_name_rot = 0,
  annotation_name_gp = gpar(fontsize = 0, fontfamily = "Helvetica"),
  annotation_legend_param = list(
    title_gp = gpar(fontsize = 8, fontface = "bold", fontfamily = "Helvetica"),
    labels_gp = gpar(fontsize = 8, fontfamily = "Helvetica")
  )
)

# col_fun <- colorRamp2(c(-2, 0, 2), viridis(n = 3))
b <- ha %v% Heatmap(
  matrix = mat,
  col = viridis(n = 100),
  cluster_rows = F,
  cluster_columns = T,
  clustering_method_columns = "ward.D2",
  show_row_names = T,
  show_column_names = F,
  name = "Enrichment (log)",
  row_split = rd$Category,
  column_split = cd[, 2, drop = FALSE],
  cluster_column_slices = FALSE,
  column_dend_height = unit(0.5, "cm"),
  show_parent_dend_line = FALSE,
  row_title = NULL,
  top_annotation = ha,
  column_title = NULL,
  row_names_gp = gpar(fontsize = 8, fontfamily = "Helvetica"),
  use_raster = FALSE,
  left_annotation = la,
  height = unit(x = 5.5, units = "cm"), width = unit(x = 3.5, units = "cm"),
  heatmap_legend_param = list(
    title = expression(bold(~ Log[2] ~ "fold-change")),
    legend_height = unit(2.5, "cm"),
    title_position = "leftcenter-rot",
    title_gp = gpar(fontsize = 8, fontfamily = "Helvetica"),
    labels_gp = gpar(fontsize = 8, fontfamily = "Helvetica")
  )
)

# pdf(file = "output/b.pdf", width = 8.5, height = 11)
draw(
  object = b,
  heatmap_legend_side = "right",
  annotation_legend_side = "right",
  merge_legends = TRUE,
  align_annotation_legend = ""
)
# dev.off()

```











# Saving results
```{r}
save(data, file = "output/repeat_elements_atac_diff_chromatin_accessibility.RData")

tab <- data.frame(rowData(data), check.names = F, stringsAsFactors = F)

data.table::fwrite(
  x = tab,
  file = "./output/repeat_elements_atac_diff_chromatin_accessibility.txt.gz",
  sep = "\t", quote = F, row.names = F, nThread = 16
)

tab.sig <- tab[abs(tab$logFC) >= 0.5 & tab$qvalue <= 0.05, ]

writexl::write_xlsx(
  x = tab.sig, path = "./output/repeat_elements_atac_diff_chromatin_accessibility_sig.xlsx",
  col_names = T, format_headers = T
)
```



# SessionInfo
```{r}
devtools::session_info()
```



```{r}
library(parallel)
bamFiles <- list.files(path = "input", pattern = "bam$", full.names = T)
names(bamFiles) <- gsub(pattern = "\\.bam", replacement = "", basename(bamFiles))

gtf <- data.frame(rtracklayer::import("input/mm10_rmsk_TE.gtf.gz"))
gtf <- gtf[gtf$class_id %in% c("LTR", "LINE", "SINE"),]

gtf$ID <- paste(gtf$class_id, gtf$family_id, sep = "_")

gtf_sp <- split(gtf, gtf$ID)


for(i in 1:length(gtf_sp)){
  rtracklayer::export(object = gtf_sp[[i]], con = paste0("./gtf/", names(gtf_sp)[i], ".gtf"), format = "gtf")
}
```

```{r}
gtf_files <- data.frame(rtracklayer::import("input/mm10_rmsk_TE.gtf.gz"))

fc_sub_family <- featureCounts(
  isPairedEnd = TRUE,
  files = bamFiles[1:2],
  annot.ext = "test.gtf",
  isGTFAnnotationFile = T,
  GTF.attrType = "class_id",
  GTF.attrType.extra = "family_id",
  GTF.featureType = "exon",
  allowMultiOverlap = T,
  fraction = T,
  countMultiMappingReads = T,
  ignoreDup = T,
  nthreads = 2
)
```


```{r}
tab <- data.table::fread(input = "output/repeat_elements_atac_diff_chromatin_accessibility.txt.gz", 
                         sep = "\t",stringsAsFactors = F)

sum(tab$class_id == "SINE")
```






```{r}
nf <- data@colData$norm.factors
names(nf) <- gsub(pattern = "input/|\\.bam", replacement = "", x = data@colData$bam.files)

rd <- c(Adult_03 = 60657376, Adult_09 = 69162383, Adult_11 = 85219624, Adult_13 = 69260772, Adult_17 = 63491836, 
  PND15_343 = 85833916, PND15_346 = 70049032, PND15_348 = 65804531, PND15_361 = 61231716, PND15_367 = 133311494, PND15_372 = 75217146)

rc <- colSums(assay(counts))
names(rc) <- gsub(pattern = "input/|\\.bam", replacement = "", x = data@colData$bam.files)

rcf <- colSums(assay(data))
names(rcf) <- gsub(pattern = "input/|\\.bam", replacement = "", x = data@colData$bam.files)

cpm <- colSums(cpm(y, log = FALSE))
names(cpm) <- gsub(pattern = "input/|\\.bam", replacement = "", x = data@colData$bam.files)

par(mfrow=c(2,2))

plot(x = rd, y = rc, main = "nReads vs RE counts", xlab = "Number of reads from bam file", ylab = "Counts of RE", 
     col = as.numeric(c(rep("1", 5), rep("2", 6))), pch = 16)
abline(rlm(rc~rd))


plot(x = rd, y = rcf, main = "nReads vs RE filtered counts", xlab = "Number of reads from bam file", ylab = "Filtered counts of RE", col = as.numeric(c(rep("1", 5), rep("2", 6))), pch = 16)
abline(rlm(rcf~rd))

plot(x = rd, y = cpm, main = "nReads vs normalization factor", xlab = "Number of reads from bam file", ylab = "Normalization factors", col = as.numeric(c(rep("1", 5), rep("2", 6))), pch = 16)
abline(rlm(cpm~rd))

plot(x = rd, y = nf, main = "nReads vs normalization factor", xlab = "Number of reads from bam file", ylab = "Normalization factors", col = as.numeric(c(rep("1", 5), rep("2", 6))), pch = 16)
abline(rlm(nf~rd))
```

